{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNYSMuJbcblQlg27Dn+Ys9U",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PStettler/Schneider-Group-DCBP/blob/main/restrictionEnyzmeMotifFinder%26Fixer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<h2>Welcome to the Restriction Enyzme Motif finder and fixer of the Schneider Group!</h2>\n",
        "\n",
        "This tool locates Restriction enzyme motifs for you and suggests changes of the sequence to fix them! Amazing no? Say no more to waking up at night from dreeming of two bands on your agarose gel!\n",
        "\n",
        "You can provide you sequence below. There is a hard-coded list of restriction enzyme motifs that we work with frequently. If you are looking for some more fancy ones, you can provide them too down below. Have fun!"
      ],
      "metadata": {
        "id": "Ni9_eMsBCvtx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sequence = \n",
        "  #please replace the example DNA sequence below with your sequence of interest, use \n",
        "  #only DNA characters (ATGCNRYSWKMBDHV) in capital letters. \n",
        "  \"\n",
        "      ATGAGCGGCCGCGCAGCGGCCGCTGATTTCTTCAACATCGTCGACGACTTCCTGAAGAAAACTTTCAGAGATGAGTCGTT\n",
        "      TCTATTCGCTGCAGTCAAGTCCAGGCAGTATTCCGAGTCTTTTCCGGGTGAACAACTCTTCTTCACCTCCCCGCCTTCGG\n",
        "      CGACGACTGACCAGCCCTCGGATGAGAGCCTTCAAGCAGGAGGAGGCGAGTTGAGGAAAACGCAAAATTTTATGTACTTC\n",
        "      TCGCCCCGTCTAATTTTCAACAGAGACGGAGGGTACCATGGGAAGGTGAAGTTGCACAGTGGTGTTCACGTGCCGCAGGT\n",
        "      GTGTCGGTTTGAGCAAGGGATTGCTGTGAACAGCGAGGGGTTGGCATCGGGTAGCGTCAAGGTTAGTGATCTTCTTGAGG\n",
        "      GAATGGAGGTAAAAGGGCGTCTCGCAGTAAACACAATTGCTCCTCCGTCAAAAGATGTGTGGTCCGTTGCAATGGATTAC\n",
        "      CAGCGTCGTGACTTCTATTCCACACTTAACTATCAGAGGAACGGATTGGGAAGCAGTGACTTGCTTGTTGATTGTGGCAC\n",
        "      GAAATTTTTTAATCTTCTTGCGGGTGCCGGATTTGAGCGGCAGAAAGTTTCGTTTCTTGAGCAACAGGATCACACCGCGC\n",
        "      AGTTGGACGTGCTTTATGCGGGGGTTGGCTTTACTGGGGTGAATTGGTCCGTTGGTGCAAAGTTAGTGAGAGCGAACGAT\n",
        "      ATGTGGAGCGCGGCACGCATTGCTTTTTATCAGCGCGTGGTGCCCGACACATCGGTGGCCTGCGCATACAATTTCGATAT\n",
        "      GGAGGAGTCACGCGTACATGTCTCGCTGGGATTTTCACAGGGTTTTCGGCTTCGCGTCCCCACGATTCTGCAGCAGCGGG\n",
        "      CGTGTGAGCAGTTGGACGTATGGACCGCCATTCTCCCGTTTGTGGGAGCATTCAAGGTGGAGAGCGGTGGCTTATGTGCG\n",
        "      GCAACCATTCGTGGCATATTTAATGGCGTGGTGCATTGGGGTTTGGTGGCGCAGAAGAACGTGTTGGTGGAGAACAGTCC\n",
        "      CATCCGTTTCGGTCTCACTCTTTCCGTGGAATCCGGTTGA\n",
        "  \n",
        "  \"\n",
        "\n",
        "user.enzyme.sites = \n",
        "  #here you can add more enzymes to the standard list that is tested \n",
        "  #(the standard list can be seen below)\n",
        "  #do it as follow: name: Motif & name2: motif2 & etc.\n",
        "  #for example: EcoRI: \"GAATTC  & SacI: GAGCTC\"\n",
        "  #write this between the partentheses below. \n",
        "  #If you don't want to add any, just leave it as is.\n",
        "  \"     \n",
        "  \n",
        "  \n",
        "  \n",
        "  \"\n",
        "\n",
        "  #Now go to Runtime, and select run everything. The results will be displayed below \n",
        "  #this box. Have fun."
      ],
      "metadata": {
        "id": "yRUClfitCwSp"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "This is the list of all presaved restriction enzyme cut sites:\n",
        "\n",
        "NotI    GCGGCCGC\n",
        "\n",
        "HindIII   AAGCTT\n",
        "\n",
        "BamHI     GGATCC\n",
        "\n",
        "XhoI      CTCGAG\n",
        "\n",
        "XbaI      TCTAGA\n",
        "\n",
        "NdeI      CATATG\n",
        "\n",
        "SacI      GAGCTC\n"
      ],
      "metadata": {
        "id": "WXy4SVhGSpGb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Standardtext f√ºr Titel\n",
        "#################################################################\n",
        "##########Restrictione enzyme site finder and fixer##############\n",
        "################Philip Stetter, 19.03.2023#######################\n",
        "#################################################################\n",
        "\n",
        "#This tool locates undesired restrictione enzyme cut sites\n",
        "#in your gene/sequence of interest and suggests modifications\n",
        "#to remove these cut sites. You are welcome.\n",
        "\n",
        "require(stringr)\n",
        "\n",
        "###########################################################################\n",
        "#hard-coding of the custumizable retrictione enzyme list\n",
        "#here, a pre-coded list ist established. The user can provide\n",
        "#additional enzyme sites that will be appended to this list.\n",
        "restriction.enzymes = c(\"NotI\"    = \"GCGGCCGC\",\n",
        "                        \"HindIII\" = \"AAGCTT\",\n",
        "                        \"BamHI\"   = \"GGATCC\",\n",
        "                        \"XhoI\"    = \"CTCGAG\",\n",
        "                        \"XbaI\"    = \"TCTAGA\",\n",
        "                        \"NdeI\"    = \"CATATG\",\n",
        "                        \"SacI\"    = \"GAGCTC\")\n",
        "\n",
        "###########################################################################\n",
        "#the used can provide more sites to be tested, write a function that can do this\n",
        "#user.enzyme.sites = \"      EcoRI: GAATTC  & sacI: GAGCTC\" #example of the user input\n",
        "#pattern has to be name: mitof, several have to be separated by &\n",
        "\n",
        "append.enzyme.list = function(current.list,user.input){\n",
        "  user.input.no.whitespace = paste(unlist(str_extract_all(user.input,\"[A-Za-z0-9:&]\")),sep=\"\",collapse=\"\")\n",
        "  namesAndSequence = unlist(str_split(user.input.no.whitespace,\"[&]\"))\n",
        "  names.new = NA\n",
        "  motif.new = NA\n",
        "  for(i in 1:length(namesAndSequence)){\n",
        "    names.new[i] = unlist(str_split(namesAndSequence[i],\":\"))[1]\n",
        "    motif.new[i] = unlist(str_split(namesAndSequence[i],\":\"))[2]\n",
        "  }\n",
        "  if(!is.na(names.new[1])){\n",
        "    new.list = c(current.list,motif.new)\n",
        "    names(new.list) = c(names(current.list),names.new)\n",
        "  }\n",
        "  return(new.list)\n",
        "}\n",
        "\n",
        "####################################################################\n",
        "\n",
        "#sequence of interest provided by the user\n",
        "\n",
        "sequence = str_replace_all(sequence,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "\n",
        "\n",
        "########################################################################\n",
        "\n",
        "\n",
        "#a function to check if the sequence has an open reading \n",
        "#frame and return the start and stop nucleotide number.\n",
        "has_ORF= function(sequence){\n",
        "  if(str_detect(sequence,\"ATG([ATGC]{3})*((TGA)|(TAA)|(TAG))\"))\n",
        "  {\n",
        "    #detect open reading frame (ORF)\n",
        "    regex = \"ATG([ATGC]{3})+(?=((TGA)|(TAA)|(TAG)))\"\n",
        "    ORF.temp = sequence\n",
        "    while(str_detect(ORF.temp,regex)){\n",
        "      #a while loop to detect the correct open reading frame\n",
        "      #starting at the first atg and ending at the closest inframe stop codon\n",
        "      ORF.temp = str_extract(ORF.temp,regex)\n",
        "      if(!str_detect(regex,\"\\\\^\")) {regex = paste0(\"^\",regex)}\n",
        "    }\n",
        "    stop.codon = str_extract(sequence,paste0(\"(?<=\",ORF.temp,\")((TGA)|(TAA)|(TAG))\"))\n",
        "    ORF = paste0(ORF.temp,stop.codon)\n",
        "    # cat(\"an open reading frame was detected, the sequence is:\\n\",ORF,\"\\n\\n\")\n",
        "    return(list(\"ORF_status\" = TRUE, \"location\" = str_locate(sequence,ORF)))\n",
        "    \n",
        "  }else{return(list(\"ORF_status\" = FALSE, \"location\" = NA))}\n",
        "}\n",
        "# has_ORF(sequence)\n",
        "if(has_ORF(sequence)$ORF_status){\n",
        "  ORF_start = has_ORF(sequence)$location[1,1]\n",
        "  ORF_end = has_ORF(sequence)$location[1,2]\n",
        "}else{cat(\"The sequence has no open reading frame, restriction enzyme sites will only be searches but not fixed.\")}\n",
        "\n",
        "\n",
        "sequence.codons = unlist(str_extract_all(substr(sequence,ORF_start,ORF_end),\"[ATGCNRYSWKMBDHV]{3}\"))\n",
        "\n",
        "######################################################################\n",
        "#load a codon table\n",
        "{ #load codon table\n",
        "  codons.1 = c(rep(\"T\",16),rep(\"C\",16),rep(\"A\",16),rep(\"G\",16))\n",
        "  codons.2 = rep(c(rep(\"T\",4),rep(\"C\",4),rep(\"A\",4),rep(\"G\",4)),4)\n",
        "  codons.3 = rep(c(\"T\",\"C\",\"A\",\"G\"),16)\n",
        "  one.letter.not = c(\"F\",\"F\",\"L\",\"L\",\"S\",\"S\",\"S\",\"S\",\"Y\",\"Y\",\"*\",\"*\",\"C\",\"C\",\"*\",\"W\",\n",
        "                     \"L\",\"L\",\"L\",\"L\",\"P\",\"P\",\"P\",\"P\",\"H\",\"H\",\"Q\",\"Q\",\"R\",\"R\",\"R\",\"R\",\n",
        "                     \"I\",\"I\",\"I\",\"M\",\"T\",\"T\",\"T\",\"T\",\"N\",\"N\",\"K\",\"K\",\"S\",\"S\",\"R\",\"R\",\n",
        "                     \"V\",\"V\",\"V\",\"V\",\"A\",\"A\",\"A\",\"A\",\"D\",\"D\",\"E\",\"E\",\"G\",\"G\",\"G\",\"G\")\n",
        "  codons = cbind(paste(codons.1,codons.2,codons.3,sep=\"\"),one.letter.not)\n",
        "}#load codon table end\n",
        "\n",
        "#load T. brucei codon usage \n",
        "{ #codon usage table of T. brucei from kazusa.org\n",
        "  # UUU F 0.56 20.5 ( 53927)  UCU S 0.16 12.6 ( 33149)  UAU Y 0.44 11.3 ( 29639)  UGU C 0.49 10.9 ( 28666)\n",
        "  # UUC F 0.44 15.9 ( 41800)  UCC S 0.16 12.5 ( 32896)  UAC Y 0.56 14.1 ( 37043)  UGC C 0.51 11.3 ( 29707)\n",
        "  # UUA L 0.10  9.8 ( 25683)  UCA S 0.17 13.5 ( 35511)  UAA * 0.36  0.7 (  1874)  UGA * 0.36  0.7 (  1875)\n",
        "  # UUG L 0.21 19.5 ( 51383)  UCG S 0.15 11.5 ( 30350)  UAG * 0.27  0.5 (  1389)  UGG W 1.00 10.9 ( 28786)\n",
        "  # \n",
        "  # CUU L 0.24 22.3 ( 58573)  CCU P 0.23 11.1 ( 29130)  CAU H 0.47 11.3 ( 29758)  CGU R 0.23 15.9 ( 41800)\n",
        "  # CUC L 0.17 15.6 ( 40909)  CCC P 0.23 11.1 ( 29151)  CAC H 0.53 13.0 ( 34152)  CGC R 0.21 14.5 ( 38107)\n",
        "  # CUA L 0.09  8.3 ( 21743)  CCA P 0.29 13.9 ( 36504)  CAA Q 0.45 16.9 ( 44506)  CGA R 0.13  9.0 ( 23569)\n",
        "  # CUG L 0.20 18.5 ( 48568)  CCG P 0.25 11.8 ( 30918)  CAG Q 0.55 21.0 ( 55275)  CGG R 0.18 12.1 ( 31803)\n",
        "  # \n",
        "  # AUU I 0.47 19.0 ( 49922)  ACU T 0.23 13.0 ( 34212)  AAU N 0.49 18.2 ( 47864)  AGU S 0.19 15.1 ( 39666)\n",
        "  # AUC I 0.29 11.6 ( 30419)  ACC T 0.21 12.1 ( 31844)  AAC N 0.51 19.2 ( 50557)  AGC S 0.17 13.6 ( 35753)\n",
        "  # AUA I 0.25 10.0 ( 26313)  ACA T 0.30 17.5 ( 45955)  AAA K 0.44 20.8 ( 54602)  AGA R 0.10  6.8 ( 17992)\n",
        "  # AUG M 1.00 23.4 ( 61545)  ACG T 0.26 14.8 ( 39063)  AAG K 0.56 26.6 ( 69988)  AGG R 0.15 10.0 ( 26246)\n",
        "  # \n",
        "  # GUU V 0.30 22.9 ( 60206)  GCU A 0.25 20.8 ( 54709)  GAU D 0.55 28.1 ( 73880)  GGU G 0.34 22.6 ( 59534)\n",
        "  # GUC V 0.15 11.4 ( 29932)  GCC A 0.22 18.3 ( 48021)  GAC D 0.45 22.7 ( 59712)  GGC G 0.22 14.9 ( 39184)\n",
        "  # GUA V 0.17 12.6 ( 33047)  GCA A 0.28 23.3 ( 61330)  GAA E 0.46 31.7 ( 83517)  GGA G 0.23 15.6 ( 41118)\n",
        "  # GUG V 0.38 28.6 ( 75313)  GCG A 0.25 20.7 ( 54480)  GAG E 0.54 38.0 ( 99928)  GGG G 0.21 13.9 ( 36525)\n",
        "  t.brucei.codon.usage.fraction = list(c(0.56,0.44),c(0.1,0.21,0.24,0.17,0.09,0.2),c(0.16,0.16,0.17,0.15,0.19,0.17),\n",
        "                                       c(0.44,0.56),c(0.36,0.28,0.36),c(0.49,0.51),c(1),c(0.23,0.23,0.29,0.25),\n",
        "                                       c(0.47,0.53),c(0.45,0.55),c(0.23,0.21,0.13,0.18,0.10,0.15),c(0.46,0.29,0.25),\n",
        "                                       c(1),c(0.23,0.21,0.3,0.26),c(0.49,0.51),c(0.44,0.56),c(0.3,0.15,0.17,0.38),\n",
        "                                       c(0.25,0.22,0.28,0.25),c(0.55,0.45),c(0.46,0.54),c(0.34,0.22,0.23,0.21))\n",
        "  t.brucei.codon.usage.codon = list(c(\"TTT\",\"TTC\"),c(\"TTA\",\"TTG\",\"CTT\",\"CTC\",\"CTA\",\"CTG\"),c(\"TCT\",\"TCC\",\"TCA\",\"TCG\",\"AGT\",\"AGC\"),\n",
        "                                    c(\"TAT\",\"TAC\"),c(\"TAA\",\"TAG\",\"TGA\"),c(\"TGT\",\"TGC\"),c(\"TGG\"),c(\"CCT\",\"CCC\",\"CCA\",\"CCG\"),\n",
        "                                    c(\"CAT\",\"CAC\"),c(\"CAA\",\"CAG\"),c(\"CGT\",\"CGC\",\"CGA\",\"CGG\",\"AGA\",\"AGG\"),c(\"ATT\",\"ATC\",\"ATA\"),\n",
        "                                    c(\"ATG\"),c(\"ACT\",\"ACC\",\"ACA\",\"ACG\"),c(\"AAT\",\"AAC\"),c(\"AAA\",\"AAG\"),c(\"GTT\",\"GTC\",\"GTA\",\"GTG\"),\n",
        "                                    c(\"GCT\",\"GCC\",\"GCA\",\"GCG\"),c(\"GAT\",\"GAC\"),c(\"GAA\",\"GAG\"),c(\"GGT\",\"GGC\",\"GGA\",\"GGG\"))\n",
        "  names(t.brucei.codon.usage.fraction) = c(\"F\",\"L\",\"S\",\"Y\",\"*\",\"C\",\"W\",\"P\",\"H\",\"Q\",\"R\",\"I\",\"M\",\"T\",\"N\",\"K\",\"V\",\"A\",\"D\",\"E\",\"G\")\n",
        "  names(t.brucei.codon.usage.codon) = c(\"F\",\"L\",\"S\",\"Y\",\"*\",\"C\",\"W\",\"P\",\"H\",\"Q\",\"R\",\"I\",\"M\",\"T\",\"N\",\"K\",\"V\",\"A\",\"D\",\"E\",\"G\")\n",
        "} #fuck me\n",
        "\n",
        "\n",
        "#############################################################################################\n",
        "\n",
        "\n",
        "#getting started, analyze the provided sequence\n",
        "#update the restriction enzyme list (if given)\n",
        "if(str_detect(user.enzyme.sites,\"[A-Za-z0-9:&]\")){\n",
        "restriction.enzymes = append.enzyme.list(restriction.enzymes,user.enzyme.sites)\n",
        "}\n",
        "\n",
        "#locate all restrictione enzyme cut sites in the user sequence\n",
        "sites.in.sequence = NA ; j=1\n",
        "sites.not.in.sequence = NA ; k=1\n",
        "for(i in 1:length(restriction.enzymes)){\n",
        "  if(str_detect(sequence,restriction.enzymes[i])){\n",
        "    sites.in.sequence[j] = restriction.enzymes[i]\n",
        "    names(sites.in.sequence)[j] = names(restriction.enzymes)[i]\n",
        "    j = j+1\n",
        "  }else{\n",
        "    sites.not.in.sequence[k] = restriction.enzymes[i]\n",
        "    names(sites.not.in.sequence)[k] = names(restriction.enzymes)[i]\n",
        "    k = k+1\n",
        "  }\n",
        "}\n",
        "if(j==1){cat(\"The provided sequence does not contain any of the tested Restriction enzyme cut sites.\",\n",
        "            \"\\n\\n The following enzymes were tested:\\n\\n\");print(data.frame(motif=restriction.enzymes))}\n",
        "\n",
        "if(!is.na(sites.in.sequence[1])){\n",
        "#locate the sites that are in the sequence\n",
        "sites.location.start = NA\n",
        "sites.location.end = NA\n",
        "sites.location.withinORF = NA\n",
        "sites.location.sequence = NA\n",
        "k = 1\n",
        "for(i in 1:length(sites.in.sequence)){\n",
        "  #count how often the site occurs\n",
        "  count = str_count(sequence,sites.in.sequence[i])\n",
        "  for(j in 1:count){\n",
        "    #retrieve all locations and check if they are within the ORF\n",
        "    sites.location.start[k] = str_locate_all(sequence,sites.in.sequence[i])[[1]][j,1]\n",
        "    sites.location.end[k] = str_locate_all(sequence,sites.in.sequence[i])[[1]][j,2]\n",
        "    sites.location.withinORF[k] = ifelse(sites.location.start[k] <= ORF_end & \n",
        "                                           sites.location.end[k] >= ORF_start,\n",
        "                                         TRUE,FALSE)\n",
        "    sites.location.sequence[k] = sites.in.sequence[i]\n",
        "    names(sites.location.start)[k] = paste0(names(sites.in.sequence)[i],\"#\",j)\n",
        "    names(sites.location.end)[k] = paste0(names(sites.in.sequence)[i],\"#\",j)\n",
        "    names(sites.location.withinORF)[k] = paste0(names(sites.in.sequence)[i],\"#\",j)\n",
        "    k = k+1\n",
        "  }\n",
        "}\n",
        "\n",
        "\n",
        "#print all results to console\n",
        "if(has_ORF(sequence)$ORF_status){\n",
        "cat(\"An open reading frame (OFR) was detected in the sequence, it starts at nucleotide \", ORF_start,\n",
        "    \" and ends at nucleotide \", ORF_end,\".\\n\\n\")\n",
        "}\n",
        "cat(\"The following restriction enzyme sites were detected in the provided sequence \\n (the # indicates how\", \n",
        "    \"many times the site was found)\\n\\n\")\n",
        "print(data.frame(\"start nucleotide\" = sites.location.start,\n",
        "                 \"end nucleotide\" = sites.location.end))\n",
        "cat(\"\\n\\n This is the list of all tested restriction enzyme cut sites:\\n\\n\")\n",
        "print(data.frame(motif=restriction.enzymes))\n",
        "\n",
        "\n",
        "\n",
        "#lets move on with fixing the sites\n",
        "has_restriction_enzyme_sites = sum(str_detect(sequence,restriction.enzymes))>0\n",
        "\n",
        "runs=0 #there is a risk of not finding a solution, se we shall use a break point\n",
        "\n",
        "set.seed(1)\n",
        "\n",
        "#fix the restriction enzyme cut sites\n",
        "if(has_ORF(sequence)$ORF_status){\n",
        "  \n",
        "  while(has_restriction_enzyme_sites | runs > 1000){\n",
        "    \n",
        "  #go through the list of detected sites\n",
        "  #and try to fix them\n",
        "  sequence.modified = sequence\n",
        "  \n",
        "  for(i in 1:length(sites.location.withinORF)){\n",
        "    if(!sites.location.withinORF[i]){cat(\"The site of \",names(sites.location.withinORF)[i],\n",
        "                                         \" was not tried to be fixed as it is outside the ORF\")\n",
        "    }else{\n",
        "      #calculate in which codons the site is\n",
        "      if(sites.location.start[i] >= ORF_start){#site must start in a codon\n",
        "        codon.site.start = ceiling(((sites.location.start[i] - ORF_start) +1)/3)\n",
        "        codon.site.end = ceiling(((sites.location.end[i] - ORF_start) +1)/3)\n",
        "        #site could end outside ORF, check and correct if neccessary\n",
        "        codon.site.end = ifelse(codon.site.end > length(sequence.codons),length(sequence.codons),codon.site.end)\n",
        "        \n",
        "      }else{#site must end in the ORF\n",
        "        codon.site.start = 1\n",
        "        codon.site.end = ceiling(((sites.location.end[i] - ORF_start) +1)/3)\n",
        "      }\n",
        "      \n",
        "      \n",
        "      sequence.modified.test = sequence.modified\n",
        "      while(sum(str_detect(substr(sequence.modified.test,sites.location.start[i],sites.location.end[i]),\n",
        "                           restriction.enzymes))>0){\n",
        "        #always re-set the test sequence to the original modified one\n",
        "        #so that unneccessary changes are not saved\n",
        "        squence.modified.test = sequence.modified\n",
        "        \n",
        "        #modify the codons in the loop until no sites are present anymore\n",
        "        #this loop will only remove one site at a time within the for loop\n",
        "        #over all sites. Since next to the changed sites new sites might appear,\n",
        "        #everything is run within another while loop.\n",
        "        #ramdomly select one codon and change it\n",
        "        codon.to.change = sample(codon.site.start:codon.site.end,1)\n",
        "        aa = codons[which(sequence.codons[codon.to.change]==codons[,1]),2]\n",
        "        new.tripplets = sample(unlist(t.brucei.codon.usage.codon[aa]),\n",
        "                               1,\n",
        "                               prob = unlist(t.brucei.codon.usage.fraction[aa]),\n",
        "                               replace = TRUE)\n",
        "        #replace the codon\n",
        "        substr(sequence.modified.test,(codon.to.change-1)*3 +1,(codon.to.change-1)*3 +1 +2) = new.tripplets\n",
        "      }\n",
        "      sequence.modified = sequence.modified.test #the site was fixed, so save and move on\n",
        "    }\n",
        "  }\n",
        "  \n",
        "  runs = runs +1\n",
        "  \n",
        "  has_restriction_enzyme_sites = sum(str_detect(sequence.modified,restriction.enzymes))>0\n",
        "  \n",
        "  }\n",
        "  \n",
        "}\n",
        "#exit due to no solution?\n",
        "if(runs>1000){cat(\"no solution to remove all sites was found. sorry.\")\n",
        "}else{\n",
        "  #pretty print the solution  \n",
        "  \n",
        "  track.of.changes = unlist(str_split(sequence.modified,\"\"))\n",
        "  track.of.changes[track.of.changes==unlist(str_split(sequence,\"\"))] = \"|\"\n",
        "  sites.on.sequence = rep(\" \",nchar(sequence))\n",
        "  for(i in 1:length(sites.location.start)){\n",
        "    sites.on.sequence[sites.location.start[i]:(sites.location.start[i]+nchar(names(sites.location.start)[i])-1)] = \n",
        "      unlist(str_split(names(sites.location.start)[i],\"\"))\n",
        "  }\n",
        "  sites.on.sequence = paste(sites.on.sequence,sep=\"\",collapse=\"\")\n",
        "  sequence = paste(sequence,sep=\"\",collapse=\"\")\n",
        "  track.of.changes = paste(track.of.changes,sep=\"\",collapse=\"\")\n",
        "  sequence.modified = paste(sequence.modified,sep=\"\",collapse=\"\")\n",
        "  \n",
        "  cat(\"\\n\\nThis is the overview of the suggested changes:\\n\",\n",
        "      \"(1: detected enzyme sites, 2: input sequence, 3: modifications to remove motifs, 4: suggested sequence)\\n\")\n",
        "  l = 1\n",
        "  continue.print = TRUE\n",
        "  while(continue.print){\n",
        "    if(l==nchar(sequence)){continue.print = FALSE}\n",
        "    else{sites.on.sequence.out = substr(sites.on.sequence,l,l+99)\n",
        "    sequence.out = substr(sequence,l,l+99)\n",
        "    track.of.changes.out = substr(track.of.changes,l,l+99)\n",
        "    sequence.modified.out = substr(sequence.modified,l,l+99)\n",
        "    print(c(sites.on.sequence.out,sequence.out,track.of.changes.out,sequence.modified.out))\n",
        "    l = l+100\n",
        "    l = ifelse(l<=nchar(sequence),l,nchar(sequence))\n",
        "    cat(\"\\n\")\n",
        "    }\n",
        "  }\n",
        "  \n",
        "  cat(\"This is the final result (to be copy and pasted):\\n\\n\")\n",
        "  print(sequence.modified)\n",
        "  \n",
        "  \n",
        "  }\n",
        "\n",
        "\n",
        "}\n",
        "\n",
        "#end of script"
      ],
      "metadata": {
        "id": "k669_QrUFV6q",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}