{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPV/PCLzG9DeDf7KkuGhf1G",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PStettler/Schneider-Group-DCBP/blob/main/sequencingResultAnalyzer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Welcome to the sequence cleaner! <br>\n",
        "Is there something more annoying than copying a sequence from somewhere which contains line breaks or numbers or even U instead of T? I don't think so. This is why you need this algorithm!"
      ],
      "metadata": {
        "id": "ob4_HIUr5Llw"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "9sHPsjtn5IId"
      },
      "outputs": [],
      "source": [
        "  #please replace the example DNA sequences below with your sequencing results as\n",
        "  #well as the sequence that should be encoded on the plasmid by now (target). Use \n",
        "  #only DNA characters (GATC + see list below). \n",
        "  #Now go to Runtime, and select run everything. The results will be displayed below \n",
        "  #this box. Have fun.\n",
        "\n",
        "  # R for A or G\n",
        "  # Y for C or T\n",
        "  # S for G or C\n",
        "  # W for A or T\n",
        "  # K for G or T\n",
        "  # M for A or C\n",
        "  # B for C, G, or T\n",
        "  # D for A, G, or T\n",
        "  # H for A, C, or T\n",
        "  # V for A, C, or G\n",
        "  # N for A, G, C, or T\n",
        "\n",
        "\n",
        "sequence_20_32 = \n",
        "\n",
        "#sequencing result with primer 20/32  \n",
        "  \n",
        "\"\n",
        "\n",
        "GACTTCAATTACACCAAAAAGTAAAATTCACAAGCTTATGAACAATGCTAATGACACTTTA\n",
        "GATTACAGTGGTAACACTTCCTCTACGTATTCTGCCGCTGTAAACGAAGAATTTTTTACTC\n",
        "CTACAAGACGTACTCATGTGTCAGTGTCCACTTCTCACGTGAGCAGCTCCGAAGCTGACTT\n",
        "CGTCACGGGATCAATGGTCGACGACGAAGAATATTACAATGAATACACTAGTGGGTATGAA\n",
        "AGCAGAGTGTATACTCATGCATCCTTCACCTACGACGAATTTTCTTCTGGGTCCCCAAATT\n",
        "CCGACGACCTGAGTTCAACGTTCATGAGCTACTATATTGATAATGTCATAGGCCAAAAGAC\n",
        "TCGCTATTTTTTTTCTGTTCTGCGAAAGCCAATCGTGTTCTTCTCGTTGATATTGCCTGGA\n",
        "TCTAACTTTGTTGGGTCCCTTGGTCGTCTGGTATACGCTGTAACTGTATCGGCGTCCAGAC\n",
        "GCACAGCGGTAAATTATTCTATGGGCGTGGAAACCCTTTGCGCCCTTCCTAAGGGGTATGG\n",
        "GCACATGCCGGAGCACTTGCAGCGCGCACGTCTTGCATTGGAGTCATTTGTTGAGTACTTC\n",
        "GCTAAAATGTTGTTGCGACTCAGTGGAGTTCCTTCGCTTGCTGATTTGCAGAAGAGAAAAT\n",
        "GGTTAGAAACGGTGCGAATTCTGAAGAGTGCAACTGGGGAGATTATATCACAGTCAGCCCA\n",
        "CCGGTCTTCAGGGCCAATGATACCGCTTGCAGTTTTCCTTCCAGTGTTTTCACTCAGTGTT\n",
        "ATTACGATAGCTCGTGGCGTGATTTCACGGCAACAAACCTATAACGAATATGAGGGATCCG\n",
        "AAAATTTATATTTTCAAGGTAGATCCGGTACCGGGCCCCCCCTCGAGGGCGGCCGATCGAG\n",
        "ATCCGAGGAGCAGAAGCTGATCTCGGAGGAGGATCTGCTGAGATCCGAGGAGCAGAAGCTG\n",
        "ATCTCGGAGGAGGATCTGCTGAGATCCGAGGAGCAGAAGCTGATCTCGGAGGAGGATCTGC\n",
        "TGTGAGATCCTGCCCATTTAGTTGGCTTTTCCCTTGTCTCGKGTCTTTTCCGTGGAAAGGT\n",
        "TCCCGGAGTAATCTGATGGCMCAGCAGGGAGGTGCGCCTGCAGGTTGGTTAGGAAGGGGGG\n",
        "ATGATGTAAAAGAAGAAA\n",
        "\n",
        "\n",
        "\n",
        "\"\n",
        "\n",
        "sequence_20_33 = \n",
        "  \n",
        "  #sequencing result with primer 20/33 \n",
        "  \n",
        "\"\n",
        "\n",
        "ACAGGGAAAGCCAACTAAATGGGCAGGATCTCACAGCAGATCCTCCTCCGAGATCAGCTTC\n",
        "TGCTCCTCGGATCTCAGCAGATCCTCCTCCGAGATCAGCTTCTGCTCCTCGGATCTCAGCA\n",
        "GATCCTCCTCCGAGATCAGCTTCTGCTCCTCGGATCTCGATCGGCCGCCCTCGAGGGGGGG\n",
        "CCCGGTACCGGATCTACCTTGAAAATATAAATTTTCGGATCCCTCATATTCGTTATAGGTT\n",
        "TGTTGCCGTGAAATCACGCCACGAGCTATCGTAATAACACTGAGTGAAAACACTGGAAGGA\n",
        "AAACTGCAAGCGGTATCATTGGCCCTGAAGACCGGTGGGCTGACTGTGATATAATCTCCCC\n",
        "AGTTGCACTCTTCAGAATTCGCACCGTTTCTAACCATTTTCTCTTCTGCAAATCAGCAAGC\n",
        "GAAGGAACTCCACTGAGTCGCAACAACATTTTAGCGAAGTACTCAACAAATGACTCCAATG\n",
        "CAAGACGTGCGCGCTGCAAGTGCTCCGGCATGTGCCCATACCCCTTAGGAAGGGCGCAAAG\n",
        "GGTTTCCACGCCCATAGAATAATTTACCGCTGTGCGTCTGGACGCCGATACAGTTACAGCG\n",
        "TATACCAGACGACCAAGGGACCCAACAAAGTTAGATCCAGGCAATATCAACGAGAAGAACA\n",
        "CGATTGGCTTTCGCAGAACAGAAAAAAAATAGCGAGTCTTTTGGCCTATGACATTATCAAT\n",
        "ATAGTAGCTCATGAACGTTGAACTCAGGTCGTCGGAATTTGGGGACCCAGAAGAAAATTCG\n",
        "TCGTAGGTGAAGGATGCATGAGTATACACTCTGCTTTCATACCCACTAGTGTATTCATTGT\n",
        "AATATTCTTCGTCGTCGACCATTGATCCCGTGACGAAGTCAGCTTCGGAGCTGCTCACGTG\n",
        "AGAAGTGGACACTGACACATGAGTACGTCTTGTAGGAGTAAAAAATTCTTCGTTTACAGCG\n",
        "GCAGAATACGTAGAGGAAGTGTTACCACTGTAATCTAAAKGCATAG\n",
        "\n",
        "\n",
        "\"\n",
        "\n",
        "\n",
        "target_gene = \n",
        "\n",
        "#the sequence of the gene of interest  \n",
        "  \n",
        "\"\n",
        "\n",
        "ATGAACAATGCTAATGACACTTTAGATTACAGTGGTAACACTTCCTCTACGTATTCTGCCGCTGTAAACGAAGAATTTTT\n",
        "TACTCCTACAAGACGTACTCATGTGTCAGTGTCCACTTCTCACGTGAGCAGCTCCGAAGCTGACTTCGTCACGGGATCAA\n",
        "TGGTCGACGACGAAGAATATTACAATGAATACACTAGTGGGTATGAAAGCAGAGTGTATACTCATGCATCCTTCACCTAC\n",
        "GACGAATTTTCTTCTGGATCCCCAAATTCCGACGACCTGAGTTCAACGTTCATGAGCTACTATATTGATAATGTCATAGG\n",
        "CCAAAAGACTCGCTATTTTTTTTCTGTTCTGCGAAAGCCAATCGTGTTCTTCTCGTTGATATTGCCTGGATCTAACTTTG\n",
        "TTGGGTCCCTTGGTCGTCTGGTATACGCTGTAACTGTATCGGCGTCCAGACGCACAGCGGTAAATTATTCTATGGGCGTG\n",
        "GAAACCCTTTGCGCCCTTCCTAAGGGGTATGGGCACATGCCGGAGCACTTGCAGCGCGCACGTCTTGCATTGCCCTCATT\n",
        "TGTTGAGTACTTCGCTAAAATGTTGTTGCGACTCAGTGGAGTTCCTTCGCTTGCTGATTTGCAGAAGAGAAAATGGTTAG\n",
        "AAACGGTGCGAATTCTGAAGAGTGCAACTGGGGAGATTATATCACAGTCAGCCCACCGGTCTTCAGGGCCAATGATACCG\n",
        "CTTGCAGTTTTCCTTCCAGTGTTTTCACTCAGTGTTATTACGATAGCTCGAGGCGTGATTTCACGGCAACAAACCTATAA\n",
        "CGAATATGAGGAGGACGAGGACGTGGCGTACGAAGCGGCGCATTCAGGTAGTTACAAGGGGGTGGCAGCTTCGCGAGGGT\n",
        "TGGATGCGCTACAGCTGTCTGGATTAGATGACTCATTCCCAAATCTAGGATCGGAAAAATCAAACGAGGTTAACTACAGT\n",
        "GGGGTGGAAACGCCGAAAGGTTCACCACCCGCCCCCGAACTTTTCCCGTCTCCTAAACCACAAGGTGAGAGACATGCCCC\n",
        "CTCTGAGACACAAACGAACCCCCAAACAGTACCAACTTCGAGGGATTTATCTCCTGATTCTTCCGCTGTCGGTGATGAAT\n",
        "TTAAATTGATAATGTCCCCCACATCAGAGGATCCGGGGCGAGCCCCTTCTGCGGCTTCCCGGGAAGCTGTTCGGCATCCT\n",
        "AAACAGTACGATCTAAGGGGTGATGTCGTAATCGCTTTGACTGGGGAGCCGTCACCTCAAGAGGTGGATGTTCTGCTTAT\n",
        "GGCGGCGGCGAAATATAAGTGTCGGAGGGTAATTTTGCCCTCTTTAGTGGCAGACACAGCGCTGGCGTCAGTGCATCCTT\n",
        "TAGTGGATGTGGTGAAGGTCAGCGATGTGATGGAACATGTCATGAGTGTTGAGGAAAGCGAAGGGCGGACTACTGTTGGA\n",
        "GTTTTTCTGCGGCCAGTAACTGATGAAACGGAGGAATTGCGGCTCTTCACACACCCTACCTCTGTAGTATACATTTTTGT\n",
        "GCCTGCGGCATTTTTGGAGGCTAATGTTATCGCACATCTGGTTGACCGTAGCGTTTATTATTCGGCATCGTTGGACGAAC\n",
        "CCTTCCCCATCAATGTGTGTCTTTATGACCGCCTTACTAAAGAGCGACAGGAGGGTGCGTGA\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\"\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title \n",
        "#packages\n",
        "require(stringr)\n",
        "require(BiocManager)\n",
        "BiocManager::install(\"Biostrings\")\n",
        "require(Biostrings)\n",
        "#functions \n",
        "reverse_complement = function(seq)\n",
        "{ #takes ATGC characters but also\n",
        "  # R for A or G\n",
        "  # Y for C or T\n",
        "  # S for G or C\n",
        "  # W for A or T\n",
        "  # K for G or T\n",
        "  # M for A or C\n",
        "  # B for C, G, or T\n",
        "  # D for A, G, or T\n",
        "  # H for A, C, or T\n",
        "  # V for A, C, or G\n",
        "  # N for A, G, C, or T\n",
        "  \n",
        "  seq = unlist(strsplit(seq,\"\"))\n",
        "  \n",
        "  #compute complement\n",
        "  seq_complement = rep(NA,length(seq))\n",
        "  seq_complement[seq==\"A\"] = \"T\"\n",
        "  seq_complement[seq==\"T\"] = \"A\"\n",
        "  seq_complement[seq==\"C\"] = \"G\"\n",
        "  seq_complement[seq==\"G\"] = \"C\"\n",
        "  seq_complement[seq==\"R\"] = \"Y\"\n",
        "  seq_complement[seq==\"Y\"] = \"R\"\n",
        "  seq_complement[seq==\"S\"] = \"S\"\n",
        "  seq_complement[seq==\"W\"] = \"W\"\n",
        "  seq_complement[seq==\"K\"] = \"M\"\n",
        "  seq_complement[seq==\"M\"] = \"K\"\n",
        "  seq_complement[seq==\"B\"] = \"V\"\n",
        "  seq_complement[seq==\"V\"] = \"B\"\n",
        "  seq_complement[seq==\"D\"] = \"H\"\n",
        "  seq_complement[seq==\"H\"] = \"D\"\n",
        "  seq_complement[seq==\"N\"] = \"N\"\n",
        "  \n",
        "  #compute reverse of complement\n",
        "  seq_reverse_complement = rep(NA,length(seq))\n",
        "  seq_reverse_complement[length(seq):1] = seq_complement[1:length(seq)]\n",
        "  \n",
        "  return(paste(seq_reverse_complement,sep=\"\",collapse=\"\"))\n",
        "}\n",
        "\n",
        "\n",
        "\n",
        "sequencing_results_analyzer = function(seq1,seq2,target)\n",
        "{\n",
        "  seq_20_32 = str_replace_all(seq1,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "  seq_20_33 = str_replace_all(seq2,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "  \n",
        "  #reverse complement of 20_33 is needed\n",
        "  seq_20_33 = reverse_complement(seq_20_33)\n",
        "  cat(\"the reverse complement of the sequencing result with primer 20/33 is: \\n\",seq_20_33,\"\\n\\n\")\n",
        "  \n",
        "  #scan for overlapping the sequence between 20_32 and 20_33 revcomp.\n",
        "  length_20_32 = nchar(seq_20_32)\n",
        "  sequence_combined = NA\n",
        "  for(i in seq(round(length_20_32/2),length_20_32,length.out = 20)[-20])\n",
        "  { #the loop is simply to check 20 parts of the 2. half of 20_32 sequence \n",
        "    #for presence in 2. half of 20_33 revcomp sequence\n",
        "    query = substr(seq_20_32,i,i+50) #value 50 randomely chosen by me\n",
        "    \n",
        "    if(str_detect(query,\"[NRYSWKMBDHV]\")){\n",
        "      sequencing_error = TRUE\n",
        "      next()\n",
        "    }else{sequencing_error = FALSE}\n",
        "    \n",
        "    if(str_detect(seq_20_33,query)){\n",
        "      sequence_combined = paste(substr(seq_20_32,1,i+50),\n",
        "                                substr(seq_20_33,str_locate(seq_20_33,query)[2]+1,nchar(seq_20_33)),\n",
        "                                sep=\"\",collapse=\"\")\n",
        "      break()\n",
        "    }\n",
        "  }\n",
        "  if(!is.na(sequence_combined)){\n",
        "    cat(\"the sequence is: \\n\",sequence_combined)\n",
        "    cat(\"the sequence was generated with the first\",i+50,\n",
        "        \"nucleotides of the 20_32 sequence \\nand last\",nchar(seq_20_33)-(str_locate(seq_20_33,query)[2]),\n",
        "        \"nucleotides of the reverse complement of the 20_32 sequence\\n\\n\")\n",
        "  }else{\n",
        "    if(sequencing_error){cat(\"no combined sequence could be generated\",\n",
        "                             \"sequencing errors were detected which could be the problem\\n\\n\")}\n",
        "    else{cat(\"please check input sequences, something unexpected went wrong.\\n\")}\n",
        "  }\n",
        "  \n",
        "  \n",
        "  #detect the open reading frame (OFR) and check it against a provided sequence\n",
        "  if(str_detect(sequence_combined,\"ATG([ATGC]{3})*((TGA)|(TAA)|(TAG))\"))\n",
        "  {\n",
        "    #detect open reading frame (ORF)\n",
        "    regex = \"ATG([ATGC]{3})+(?=((TGA)|(TAA)|(TAG)))\"\n",
        "    ORF = sequence_combined\n",
        "    while(str_detect(ORF,regex)){\n",
        "      #a while loop to detect the correct open reading frame\n",
        "      #starting at the first atg and ending at the closest inframe stop codon\n",
        "      ORF = str_extract(ORF,regex)\n",
        "      if(!str_detect(regex,\"\\\\^\")) {regex = paste0(\"^\",regex)}\n",
        "    }\n",
        "    cat(\"an open reading frame was detected, the sequence is:\\n\",ORF,\"\\n\\n\")\n",
        "    \n",
        "    target = str_replace_all(target,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "    \n",
        "    alignment = pairwiseAlignment(ORF,target)\n",
        "\n",
        "    #print this a little nicer into the console\n",
        "    pattern = unlist(strsplit(as.character(alignedPattern(alignment)),\"\"))\n",
        "    subject = unlist(strsplit(as.character(alignedSubject(alignment)),\"\"))\n",
        "\n",
        "    matching = rep(\" \",length(pattern))\n",
        "\n",
        "    matching[which(pattern==subject)] = \"|\"\n",
        "    matching = paste(matching,sep=\"\",collapse=\"\")\n",
        "    \n",
        "    cat(\"the sequence alignment to the target sequence looks as follows:\\n\")\n",
        "    print(as.character(c(alignedPattern(alignment), matching, alignedSubject(alignment))),width=100)\n",
        "\n",
        "  }\n",
        "  \n",
        "}\n",
        "\n",
        "sequencing_results_analyzer(seq1 = sequence_20_32,seq2 = sequence_20_33,target = target_gene)\n",
        "\n"
      ],
      "metadata": {
        "id": "wD_Y3H525M4d"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}