{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPAerM2wzSp3fhcxHeaLLT2",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PStettler/Schneider-Group-DCBP/blob/main/sequencingResultAnalyzer2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Welcome to the sequencing results analyzer! <br>\n",
        "This alogrithm helps you to analyze your sequencing results from microsynth in no time. Simply copy paste the optained results from directly below and (optionally) provide the sequence of your gene that was cloned into the plasmid to check for potentional errors. In case you cloned a RNAi plasmid, you have to indicate this in the sequence from the 20/32 primer by adding \"rnai\" to the sequence in any place (see below). \n",
        "\n",
        "Runtime information: the alogrithm makes use of the needlemann-wunsch alogrithm for sequence alignments. Since I am too lazy to implement this myself, an implementation by Biostrings is used. This package has to be installed by the google server first which is why the run will take probably like 30 seconds. May you fogive me. \n"
      ],
      "metadata": {
        "id": "ob4_HIUr5Llw"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "9sHPsjtn5IId"
      },
      "outputs": [],
      "source": [
        "  #please replace the example DNA sequences below with your sequencing results as\n",
        "  #well as the sequence that should be encoded on the plasmid by now (target). Use \n",
        "  #only DNA characters (GATC + see list below). \n",
        "  #Now go to Runtime, and select run everything. The results will be displayed below \n",
        "  #this box. Have fun.\n",
        "\n",
        "  # R for A or G\n",
        "  # Y for C or T\n",
        "  # S for G or C\n",
        "  # W for A or T\n",
        "  # K for G or T\n",
        "  # M for A or C\n",
        "  # B for C, G, or T\n",
        "  # D for A, G, or T\n",
        "  # H for A, C, or T\n",
        "  # V for A, C, or G\n",
        "  # N for A, G, C, or T\n",
        "\n",
        "\n",
        "sequence_20_32 = \n",
        "\n",
        "#sequence of sequencing result with primer 20/32  \n",
        "  \n",
        "\"\n",
        "\n",
        "GACTTCAATTACACCAAAAAGTAAAATTCACAAGCTTATGAACAATGCTAATGACACTTTA\n",
        "GATTACAGTGGTAACACTTCCTCTACGTATTCTGCCGCTGTAAACGAAGAATTTTTTACTC\n",
        "CTACAAGACGTACTCATGTGTCAGTGTCCACTTCTCACGTGAGCAGCTCCGAAGCTGACTT\n",
        "CGTCACGGGATCAATGGTCGACGACGAAGAATATTACAATGAATACACTAGTGGGTATGAA\n",
        "AGCAGAGTGTATACTCATGCATCCTTCACCTACGACGAATTTTCTTCTGGGTCCCCAAATT\n",
        "CCGACGACCTGAGTTCAACGTTCATGAGCTACTATATTGATAATGTCATAGGCCAAAAGAC\n",
        "TCGCTATTTTTTTTCTGTTCTGCGAAAGCCAATCGTGTTCTTCTCGTTGATATTGCCTGGA\n",
        "TCTAACTTTGTTGGGTCCCTTGGTCGTCTGGTATACGCTGTAACTGTATCGGCGTCCAGAC\n",
        "GCACAGCGGTAAATTATTCTATGGGCGTGGAAACCCTTTGCGCCCTTCCTAAGGGGTATGG\n",
        "GCACATGCCGGAGCACTTGCAGCGCGCACGTCTTGCATTGGAGTCATTTGTTGAGTACTTC\n",
        "GCTAAAATGTTGTTGCGACTCAGTGGAGTTCCTTCGCTTGCTGATTTGCAGAAGAGAAAAT\n",
        "GGTTAGAAACGGTGCGAATTCTGAAGAGTGCAACTGGGGAGATTATATCACAGTCAGCCCA\n",
        "CCGGTCTTCAGGGCCAATGATACCGCTTGCAGTTTTCCTTCCAGTGTTTTCACTCAGTGTT\n",
        "ATTACGATAGCTCGTGGCGTGATTTCACGGCAACAAACCTATAACGAATATGAGGGATCCG\n",
        "AAAATTTATATTTTCAAGGTAGATCCGGTACCGGGCCCCCCCTCGAGGGCGGCCGATCGAG\n",
        "ATCCGAGGAGCAGAAGCTGATCTCGGAGGAGGATCTGCTGAGATCCGAGGAGCAGAAGCTG\n",
        "ATCTCGGAGGAGGATCTGCTGAGATCCGAGGAGCAGAAGCTGATCTCGGAGGAGGATCTGC\n",
        "TGTGAGATCCTGCCCATTTAGTTGGCTTTTCCCTTGTCTCGKGTCTTTTCCGTGGAAAGGT\n",
        "TCCCGGAGTAATCTGATGGCMCAGCAGGGAGGTGCGCCTGCAGGTTGGTTAGGAAGGGGGG\n",
        "ATGATGTAAAAGAAGAAA\n",
        "\n",
        "\n",
        "\n",
        "\"\n",
        "\n",
        "sequence_20_33 = \n",
        "  \n",
        "  #sequence of sequencing result with primer 20/33 \n",
        "  #!!! if you plasmid is a RNAi plamsid you have to\n",
        "  #write \"rnai\" (without the quotation marks) to the sequence\n",
        "  #e.g. like that: \"  rnai  ATGGGAACCT...\"\n",
        "  \n",
        "\"\n",
        "\n",
        "ACAGGGAAAGCCAACTAAATGGGCAGGATCTCACAGCAGATCCTCCTCCGAGATCAGCTTC\n",
        "TGCTCCTCGGATCTCAGCAGATCCTCCTCCGAGATCAGCTTCTGCTCCTCGGATCTCAGCA\n",
        "GATCCTCCTCCGAGATCAGCTTCTGCTCCTCGGATCTCGATCGGCCGCCCTCGAGGGGGGG\n",
        "CCCGGTACCGGATCTACCTTGAAAATATAAATTTTCGGATCCCTCATATTCGTTATAGGTT\n",
        "TGTTGCCGTGAAATCACGCCACGAGCTATCGTAATAACACTGAGTGAAAACACTGGAAGGA\n",
        "AAACTGCAAGCGGTATCATTGGCCCTGAAGACCGGTGGGCTGACTGTGATATAATCTCCCC\n",
        "AGTTGCACTCTTCAGAATTCGCACCGTTTCTAACCATTTTCTCTTCTGCAAATCAGCAAGC\n",
        "GAAGGAACTCCACTGAGTCGCAACAACATTTTAGCGAAGTACTCAACAAATGACTCCAATG\n",
        "CAAGACGTGCGCGCTGCAAGTGCTCCGGCATGTGCCCATACCCCTTAGGAAGGGCGCAAAG\n",
        "GGTTTCCACGCCCATAGAATAATTTACCGCTGTGCGTCTGGACGCCGATACAGTTACAGCG\n",
        "TATACCAGACGACCAAGGGACCCAACAAAGTTAGATCCAGGCAATATCAACGAGAAGAACA\n",
        "CGATTGGCTTTCGCAGAACAGAAAAAAAATAGCGAGTCTTTTGGCCTATGACATTATCAAT\n",
        "ATAGTAGCTCATGAACGTTGAACTCAGGTCGTCGGAATTTGGGGACCCAGAAGAAAATTCG\n",
        "TCGTAGGTGAAGGATGCATGAGTATACACTCTGCTTTCATACCCACTAGTGTATTCATTGT\n",
        "AATATTCTTCGTCGTCGACCATTGATCCCGTGACGAAGTCAGCTTCGGAGCTGCTCACGTG\n",
        "AGAAGTGGACACTGACACATGAGTACGTCTTGTAGGAGTAAAAAATTCTTCGTTTACAGCG\n",
        "GCAGAATACGTAGAGGAAGTGTTACCACTGTAATCTAAAKGCATAG\n",
        "\n",
        "\n",
        "\"\n",
        "\n",
        "\n",
        "target_gene = \n",
        "\n",
        "#the sequence of the gene of interest. This is the gene that\n",
        "#was cloned into the plasmid. This is optional. If you don't need\n",
        "#a sequence analysis simply delete the example sequence below \n",
        "#but leave the quotation marks there. thx  \n",
        "  \n",
        "\"\n",
        "\n",
        "ATGAACAATGCTAATGACACTTTAGATTACAGTGGTAACACTTCCTCTACGTATTCTGCCGCTGTAAACGAAGAATTTTT\n",
        "TACTCCTACAAGACGTACTCATGTGTCAGTGTCCACTTCTCACGTGAGCAGCTCCGAAGCTGACTTCGTCACGGGATCAA\n",
        "TGGTCGACGACGAAGAATATTACAATGAATACACTAGTGGGTATGAAAGCAGAGTGTATACTCATGCATCCTTCACCTAC\n",
        "GACGAATTTTCTTCTGGATCCCCAAATTCCGACGACCTGAGTTCAACGTTCATGAGCTACTATATTGATAATGTCATAGG\n",
        "CCAAAAGACTCGCTATTTTTTTTCTGTTCTGCGAAAGCCAATCGTGTTCTTCTCGTTGATATTGCCTGGATCTAACTTTG\n",
        "TTGGGTCCCTTGGTCGTCTGGTATACGCTGTAACTGTATCGGCGTCCAGACGCACAGCGGTAAATTATTCTATGGGCGTG\n",
        "GAAACCCTTTGCGCCCTTCCTAAGGGGTATGGGCACATGCCGGAGCACTTGCAGCGCGCACGTCTTGCATTGCCCTCATT\n",
        "TGTTGAGTACTTCGCTAAAATGTTGTTGCGACTCAGTGGAGTTCCTTCGCTTGCTGATTTGCAGAAGAGAAAATGGTTAG\n",
        "AAACGGTGCGAATTCTGAAGAGTGCAACTGGGGAGATTATATCACAGTCAGCCCACCGGTCTTCAGGGCCAATGATACCG\n",
        "CTTGCAGTTTTCCTTCCAGTGTTTTCACTCAGTGTTATTACGATAGCTCGAGGCGTGATTTCACGGCAACAAACCTATAA\n",
        "CGAATATGAGGAGGACGAGGACGTGGCGTACGAAGCGGCGCATTCAGGTAGTTACAAGGGGGTGGCAGCTTCGCGAGGGT\n",
        "TGGATGCGCTACAGCTGTCTGGATTAGATGACTCATTCCCAAATCTAGGATCGGAAAAATCAAACGAGGTTAACTACAGT\n",
        "GGGGTGGAAACGCCGAAAGGTTCACCACCCGCCCCCGAACTTTTCCCGTCTCCTAAACCACAAGGTGAGAGACATGCCCC\n",
        "CTCTGAGACACAAACGAACCCCCAAACAGTACCAACTTCGAGGGATTTATCTCCTGATTCTTCCGCTGTCGGTGATGAAT\n",
        "TTAAATTGATAATGTCCCCCACATCAGAGGATCCGGGGCGAGCCCCTTCTGCGGCTTCCCGGGAAGCTGTTCGGCATCCT\n",
        "AAACAGTACGATCTAAGGGGTGATGTCGTAATCGCTTTGACTGGGGAGCCGTCACCTCAAGAGGTGGATGTTCTGCTTAT\n",
        "GGCGGCGGCGAAATATAAGTGTCGGAGGGTAATTTTGCCCTCTTTAGTGGCAGACACAGCGCTGGCGTCAGTGCATCCTT\n",
        "TAGTGGATGTGGTGAAGGTCAGCGATGTGATGGAACATGTCATGAGTGTTGAGGAAAGCGAAGGGCGGACTACTGTTGGA\n",
        "GTTTTTCTGCGGCCAGTAACTGATGAAACGGAGGAATTGCGGCTCTTCACACACCCTACCTCTGTAGTATACATTTTTGT\n",
        "GCCTGCGGCATTTTTGGAGGCTAATGTTATCGCACATCTGGTTGACCGTAGCGTTTATTATTCGGCATCGTTGGACGAAC\n",
        "CCTTCCCCATCAATGTGTGTCTTTATGACCGCCTTACTAAAGAGCGACAGGAGGGTGCGTGA\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\"\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title \n",
        "#packages\n",
        "require(stringr)\n",
        "#functions \n",
        "reverse_complement = function(seq)\n",
        "{ #takes ATGC characters but also\n",
        "  # R for A or G\n",
        "  # Y for C or T\n",
        "  # S for G or C\n",
        "  # W for A or T\n",
        "  # K for G or T\n",
        "  # M for A or C\n",
        "  # B for C, G, or T\n",
        "  # D for A, G, or T\n",
        "  # H for A, C, or T\n",
        "  # V for A, C, or G\n",
        "  # N for A, G, C, or T\n",
        "  \n",
        "  seq = unlist(strsplit(seq,\"\"))\n",
        "  \n",
        "  #compute complement\n",
        "  seq_complement = rep(NA,length(seq))\n",
        "  seq_complement[seq==\"A\"] = \"T\"\n",
        "  seq_complement[seq==\"T\"] = \"A\"\n",
        "  seq_complement[seq==\"C\"] = \"G\"\n",
        "  seq_complement[seq==\"G\"] = \"C\"\n",
        "  seq_complement[seq==\"R\"] = \"Y\"\n",
        "  seq_complement[seq==\"Y\"] = \"R\"\n",
        "  seq_complement[seq==\"S\"] = \"S\"\n",
        "  seq_complement[seq==\"W\"] = \"W\"\n",
        "  seq_complement[seq==\"K\"] = \"M\"\n",
        "  seq_complement[seq==\"M\"] = \"K\"\n",
        "  seq_complement[seq==\"B\"] = \"V\"\n",
        "  seq_complement[seq==\"V\"] = \"B\"\n",
        "  seq_complement[seq==\"D\"] = \"H\"\n",
        "  seq_complement[seq==\"H\"] = \"D\"\n",
        "  seq_complement[seq==\"N\"] = \"N\"\n",
        "  \n",
        "  #compute reverse of complement\n",
        "  seq_reverse_complement = rep(NA,length(seq))\n",
        "  seq_reverse_complement[length(seq):1] = seq_complement[1:length(seq)]\n",
        "  \n",
        "  return(paste(seq_reverse_complement,sep=\"\",collapse=\"\"))\n",
        "}\n",
        "\n",
        "#needlemann wunsch algorithm implemented by juliuskittler, thx man! \n",
        "#https://gist.github.com/juliuskittler/ed53696ac1e590b413aac2dddf0457f6\n",
        "\n",
        "# References: https://de.wikipedia.org/wiki/Needleman-Wunsch-Algorithmus\n",
        "\n",
        "needlemanWunsch = function(seq1, seq2, gap=-1, mismatch=-1, match=0){\n",
        "  \n",
        "  # Stop conditions\n",
        "  stopifnot(gap <= 0) # check if penalty negative\n",
        "  stopifnot(mismatch <= 0)  # check if penalty negative\n",
        "  stopifnot(match >= 0)  # check if score positive\n",
        "  \n",
        "  # Initialize col and rownames for matrices\n",
        "  len1 = nchar(seq1); len2 = nchar(seq2) # Save number of chars in each sequence\n",
        "  seq1 = unlist(strsplit(seq1, split = \"\")) # convert seq to character vector\n",
        "  seq2 = unlist(strsplit(seq2, split = \"\")) # convert seq to character vector\n",
        "  \n",
        "  # Initialize matrix M (for scores)\n",
        "  M = matrix(0, nrow = len1 + 1, ncol = len2 + 1) # Initialize matrix\n",
        "  rownames(M) = c(\"-\", seq1) # assign seq chars to matrix names\n",
        "  colnames(M) = c(\"-\", seq2) # assign seq chars to matrix names\n",
        "  M[1, ] = cumsum(c(0, rep(gap, len2))) # Fill 1st row with gap penalites\n",
        "  M[, 1] = cumsum(c(0, rep(gap, len1))) # Fill 1st col with gap penalites\n",
        "  \n",
        "  # Initialize matrix D (for directions)\n",
        "  D = matrix(0, nrow = len1 + 1, ncol = len2 + 1) # Initialize matrix\n",
        "  rownames(D) = c(\"-\", seq1) # assign seq chars to matrix names\n",
        "  colnames(D) = c(\"-\", seq2) # assign seq chars to matrix names\n",
        "  D[1, ] = rep(\"hor\") # Fill 1st row with \"hor\" for horizontal moves\n",
        "  D[, 1] = rep(\"ver\") # Fill 1st col with \"ver\" for vertical moves\n",
        "  type = c(\"dia\", \"hor\", \"ver\") # Lookup vector\n",
        "  \n",
        "  # Compute scores and save moves\n",
        "  for (i in 2:(len1 + 1)){# for every (initially zero) row\n",
        "    for (j in 2:(len2 + 1)){# for every (initially zero) col\n",
        "      hor = M[i, j - 1] + gap # horizontal move = gap for seq1\n",
        "      ver = M[i - 1, j] + gap # vertical move = gap for seq2 \n",
        "      dia = ifelse(rownames(M)[i] == colnames(M)[j], # diagonal = ifelse(chars equal, match, mismatch) \n",
        "                   M[i - 1, j - 1] + match, \n",
        "                   M[i - 1, j - 1] + mismatch)\n",
        "      M[i, j] = max(dia, hor, ver) # Save current (best) score in M\n",
        "      D[i, j] = type[which.max(c(dia, hor, ver))] # Save direction of move in D\n",
        "    }\n",
        "  } \n",
        "  \n",
        "  # Backtracing\n",
        "  align1 = c(); align2 = c() # Note: length of final alignments is unknown at this point\n",
        "  \n",
        "  while(i > 1 && j > 1){\n",
        "    \n",
        "    if(D[i, j] == \"dia\") {\n",
        "      align1 = c(rownames(M)[i], align1)\n",
        "      align2 = c(colnames(M)[j], align2)\n",
        "      j = j - 1; i = i - 1  # update indices\n",
        "    } else if (D[i, j] == \"ver\") {\n",
        "      align1 = c(rownames(M)[i], align1)\n",
        "      align2 = c(\"-\", align2) # vertical movement = gap for seq2\n",
        "      i = i - 1 # update indices\n",
        "    } else if (D[i, j] == \"hor\") {\n",
        "      align1 = c(\"-\", align1) # horizontal movement = gap for seq1\n",
        "      align2 = c(colnames(M)[j], align2) \n",
        "      j = j - 1 # update indices\n",
        "    } \n",
        "    \n",
        "  }\n",
        "  \n",
        "  # Prepare output\n",
        "  return(list(aligned_seqs = matrix(c(align1, align2), byrow = TRUE, nrow = 2),\n",
        "              score = M[nrow(M), ncol(M)], score_matrix = M, movement_matrix = D))\n",
        "  \n",
        "}\n",
        "\n",
        "#######################################################################################################\n",
        "\n",
        "sequencing_results_analyzer = function(seq1,seq2,target)\n",
        "{ #this algorithm takes sequencing results form a forward primer (20/32 on our plasmids)\n",
        "  #and a reverse primer (20/33 on our plasmids) which will overlapp in the middle.\n",
        "  #the two sequences are combined to the final sequence which is encoded on the plasmid\n",
        "  #and this plasmid is aligned against the gene that was cloned into the plasmid (target)\n",
        "  #to check for sequence errors.\n",
        "\n",
        "  seq_20_32 = str_replace_all(seq1,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "  seq_20_33 = str_replace_all(seq2,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "\n",
        "  #break is the sequences are too short for analysis\n",
        "  if(nchar(seq_20_32)<101 | nchar(seq_20_33)<101) {cat(\"one or both provided sequecing results are too short. \\n They are 20/33:\\n\",\n",
        "                                                        seq_20_32,\"\\nand 20/33:\\n\",seq_20_33)\n",
        "                                                        break()}\n",
        "  \n",
        "  #reverse complement of 20_33 is needed\n",
        "  seq_20_33 = reverse_complement(seq_20_33)\n",
        "  cat(\"the reverse complement of the sequencing result with primer 20/33 is: \\n\",seq_20_33,\"\\n\\n\")\n",
        "  \n",
        "  #scan for overlapp in the sequences between 20_32 and 20_33 revcomp.\n",
        "  length_20_32 = nchar(seq_20_32)\n",
        "  sequence_combined = NA\n",
        "  for(i in seq(round(length_20_32/2),length_20_32,length.out = 20)[-20])\n",
        "  { #the loop is simply to check 20 parts of the 2. half of 20_32 sequence \n",
        "    #for presence in the 20_33 revcomp sequence\n",
        "    query = substr(seq_20_32,i,i+50) #value 50 randomely chosen by me\n",
        "    \n",
        "    if(str_detect(query,\"[NRYSWKMBDHV]\")){\n",
        "      #if sequence has sequencing error don't use it,\n",
        "      #go to next one\n",
        "      sequencing_error = TRUE\n",
        "      next()\n",
        "    }else{sequencing_error = FALSE}\n",
        "    \n",
        "    if(str_detect(seq_20_33,query)){\n",
        "      sequence_combined = paste(substr(seq_20_32,1,i+50),\n",
        "                                substr(seq_20_33,str_locate(seq_20_33,query)[2]+1,nchar(seq_20_33)),\n",
        "                                sep=\"\",collapse=\"\")\n",
        "      break()\n",
        "    }\n",
        "  }\n",
        "  if(!is.na(sequence_combined)){\n",
        "    #print the combined sequence encoded on the plasmid\n",
        "    cat(\"the sequence is: \\n\",sequence_combined)\n",
        "    cat(\"the sequence was generated with the first\",i+50,\n",
        "        \"nucleotides of the 20_32 sequence \\nand last\",nchar(seq_20_33)-(str_locate(seq_20_33,query)[2]),\n",
        "        \"nucleotides of the reverse complement of the 20_32 sequence\\n\\n\")\n",
        "  }else{\n",
        "    #if no combined sequence could be generated, report this together with\n",
        "    #help of what might be the problem.\n",
        "    if(sequencing_error){cat(\"no combined sequence could be generated\",\n",
        "                             \"sequencing errors were detected which could be the problem\\n\\n\")}\n",
        "    else{cat(\"please check input sequences, something unexpected went wrong.\\n\")}\n",
        "  }\n",
        "  \n",
        "  #analysis of the open reading frame in the combined sequence against \n",
        "  #the gene that was cloned into the plasmid. \n",
        "  \n",
        "  #detect the open reading frame (OFR) and check it against a provided sequence\n",
        "  if(str_detect(sequence_combined,\"ATG([ATGC]{3})*((TGA)|(TAA)|(TAG))\"))\n",
        "  {\n",
        "    #detect open reading frame (ORF)\n",
        "    regex = \"ATG([ATGC]{3})+(?=((TGA)|(TAA)|(TAG)))\"\n",
        "    ORF = sequence_combined\n",
        "    while(str_detect(ORF,regex)){\n",
        "      #a while loop to detect the correct open reading frame\n",
        "      #starting at the first atg and ending at the closest inframe stop codon\n",
        "      ORF = str_extract(ORF,regex)\n",
        "      if(!str_detect(regex,\"\\\\^\")) {regex = paste0(\"^\",regex)}\n",
        "    }\n",
        "    cat(\"an open reading frame was detected, the sequence is:\\n\",ORF,\"\\n\\n\")\n",
        "    \n",
        "    #prepare target sequence for alignment\n",
        "    target = str_replace_all(target,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "    \n",
        "    alignment = needlemanWunsch(target,ORF)$aligned_seqs\n",
        "\n",
        "    #print this a little nicer into the console\n",
        "    pattern = alignment[1,]\n",
        "    subject = alignment[2,]\n",
        "\n",
        "    matching = rep(\" \",length(pattern))\n",
        "\n",
        "    matching[which(pattern==subject)] = \"|\"\n",
        "\n",
        "    #past them back to a string\n",
        "    matching = paste(matching,sep=\"\",collapse=\"\")\n",
        "    pattern = paste(pattern,sep=\"\",collapse=\"\")\n",
        "    subject = paste(subject,sep=\"\",collapse=\"\")\n",
        "\n",
        "    #prepare for nice console print out\n",
        "    alignment.out = as.character(c(pattern, matching, subject))\n",
        "    names(alignment.out) = c(\"target gene      \",\"-----------------\",\"sequencing result\")\n",
        "    \n",
        "    cat(\"the sequence alignment to the target sequence looks as follows:\\n\")\n",
        "    print(alignment.out)\n",
        "\n",
        "  }else{cat(\"no open reading frame could be detected in the sequence.\")}\n",
        "  \n",
        "}\n",
        "\n",
        "sequencing_results_analyzer(seq1 = sequence_20_32,seq2 = sequence_20_33,target = target_gene)\n",
        "\n"
      ],
      "metadata": {
        "id": "wD_Y3H525M4d"
      },
      "execution_count": 6,
      "outputs": []
    }
  ]
}