{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOF06xO44BN0ulyqM7WokUC",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PStettler/Schneider-Group-DCBP/blob/main/sequencingResultAnalyzer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<h3>Welcome to the sequencing results analyzer!</h3> <br>\n",
        "\n",
        "This alogrithm helps you to analyze your sequencing results from microsynth in no time. Simply copy paste the optained sequencing results directly below and (optionally) provide the sequence of your gene that was cloned into the plasmid to check for potentional errors. In case you cloned an RNAi plasmid, you have to indicate this in the sequence from the reverse primer by adding \"rnai\" to the sequence in any place (see below). \n",
        "\n",
        "Runtime information: the alogrithm makes use of the Needleman Wunsch alogrithm for sequence alignments. Since I am too lazy to implement this myself, an implementation by NameNeedle is used. This package has to be installed by the google server first which is why the run will take probably like 10 seconds. May you fogive me. \n"
      ],
      "metadata": {
        "id": "ob4_HIUr5Llw"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "9sHPsjtn5IId"
      },
      "outputs": [],
      "source": [
        "  #please replace the example DNA sequences below with your sequencing results as\n",
        "  #well as the sequence that should be encoded on the plasmid by now (target). Use \n",
        "  #only DNA characters (GATC + see list below). \n",
        "  #Now go to Runtime, and select run everything. The results will be displayed below \n",
        "  #this box. Have fun.\n",
        "\n",
        "  # R for A or G\n",
        "  # Y for C or T\n",
        "  # S for G or C\n",
        "  # W for A or T\n",
        "  # K for G or T\n",
        "  # M for A or C\n",
        "  # B for C, G, or T\n",
        "  # D for A, G, or T\n",
        "  # H for A, C, or T\n",
        "  # V for A, C, or G\n",
        "  # N for A, G, C, or T\n",
        "\n",
        "\n",
        "sequence_forward_primer = \n",
        "\n",
        "#sequence of sequencing result with forward primer  \n",
        "  \n",
        "\"\n",
        "\n",
        "ACTTCATTACACCAAAAAGTAAAATTCACAAGCTTATGTTTTGTTTCTTCCAAAACCGCTGGGAGTCGGGTCTTG\n",
        "GTGCCCATCTAACCGAGGAACTCATTAGTAGCGCATTCAGCGATTTTGCGCCTTACGCGGGGGCGGAGCAGCAAC\n",
        "AGACACGTTTCGTCCGCCGCCAAAGCGATGGGACAACCACCAGAGTTGCTTATGGGGAAGATGGGGGGTGGTCGG\n",
        "TAAGCATGGGAAATCCTTACGCTGGCGGCATATCTTTTGCAACGACACCAGATTTACTCTGTGATATGCAACTGC\n",
        "GCATACCACTGAAGTTGGCTAGGGATGTTGTGGAGCAGAATAGCGAAAGGGGGATCATTGGCAGTGCCGACAACA\n",
        "GTGAGGTGCTTCCAGTTTCCAGAATCTCCCTAACACCCATAACCACACCCTACGATAGTCGAAGTCCTTTTCAGT\n",
        "ATTCTGTGCCCCCTCCACTTGGTCCCACGGGTTGCTTTGGCATGGTGGAACTTTTGTACTCCTACGTGACACATG\n",
        "TGTATGGCGCTGGTTTACGGGTTGGACCGTTGTCCCTTCGAGCGACGGCTTCCCCGGGAGGCATGCCCTCAATAG\n",
        "ACAATTCGTATGTGAGTGCTTTGGGCCAAACTGCCTGTCTCGTAGACGGTTCGTGGTGCGCTCCGTTGAGTGCCT\n",
        "TTCCGTCGTGTAACCTTACAGTGGCGCTGCAGCACCCCTTTATGGCCGTTGAGGGTGATACGCTTGCAGCAATTT\n",
        "GTTTGGAGAGTTCGGCGCCAAAAACACCAGAGGGACAAGGGTGCAATAATTTTCGTCTTCTTGTTGTGCAGACGC\n",
        "TTGCAACAAGTAATAGGCAAAACAAGCAGAATCCCAGTGATTCTCACTCGCATAGCTTCAGTTCGGTTCTCAGGG\n",
        "GCGACAAAAACTGGCGTGAGCTAGGCACACTCGTTGCTGTACAGAAATCATTTCGCAACCAACGCTTTCGACTCA\n",
        "GTGCTGCTGCTATGATACACGACCATGAATTAGACTATGAAACCGCAGCTTTGGCGGATGTAACAGCCATATTTT\n",
        "CTAGACCAACGGTACTGGCGGTTGGGTTTCAACAAAGCCCGCCGTGTTGCCATGGGTATAGCGACAAAATACTTA\n",
        "ATGGACTTACCCTTAMTTGGGT\n",
        "\n",
        "\n",
        "\"\n",
        "\n",
        "sequence_reverse_primer = \n",
        "  \n",
        "  #sequence of sequencing result with reverse primer \n",
        "  #!!! if your plasmid is an RNAi plamsid you have to\n",
        "  #write \"rnai\" (without the quotation marks) to the sequence\n",
        "  #e.g. like that: \"  rnai  ATGGGAACCT...\"\n",
        "\n",
        "  #!!!for non-RNAi plasmids this sequence is optional.\n",
        "  #if only a forward primer sequencing was made, just delete the\n",
        "  #example sequence but leave the quotation marks, thx.\n",
        "  \n",
        "\"\n",
        "\n",
        "CAAGGGAAAGCCAACTAAATGGGCAGGATCTCACAGCAGATCCTCCTCCGAGATCAGCTTCTGCTCCTCGGATCT\n",
        "CAGCAGATCCTCCTCCGAGATCAGCTTCTGCTCCTCGGATCTCAGCAGATCCTCCTCCGAGATCAGCTTCTGCTC\n",
        "CTCGGATCTCGATCGGCCGCCCTCGAGGGGGGGCCCGGTACCGGATCTACCTTGAAAATATAAATTTTCGGATCC\n",
        "AAGGGCGGCTTCAAGTCCAAACCGCAAGTCCCCCGCTGTGTAATGGACACCCAATGTAAGGGTAAGTCCATTAAG\n",
        "TATTTCTGTCGCTATACCCATGGCAACACGGCGGGCTTTGTTGAAACCCAACCGCAGTACCGTTGGTCTAGAAAA\n",
        "TATGGCTGTTACATCCGCCAAAGCTGCGGTTTCATAGTCTAATTCATGGTCGTGTATCATAGCAGCAGCACTGAG\n",
        "TCGAAAGCGTTGGTTGCGAAATGATTTCTGTACAGCAACGAGTGTGCCTAGCTCACGCCAGTTTTTGTCGCCCCT\n",
        "GAGAACCGAACTGAAGCTATGCGAGTGAGAATCACTGGGATTCTGCTTGTTTTGCCTATTACTTGTTGCAAGCGT\n",
        "CTGCACAACAAGAAGACGAAAATTATTGCACCCTTGTCCCTCTGGTGTTTTTGGCGCCGAACTCTCCAAACAAAT\n",
        "TGCTGCAAGCGTATCACCCTCAACGGCCATAAAGGGGTGCTGCAGCGCCACTGTAAGGTTACACGACGGAAAGGC\n",
        "ACTCAACGGAGCGCACCACGAACCGTCTACGAGACAGGCAGTTTGGCCCAAAGCACTCACATACGAATTGTCTAT\n",
        "TGAGGGCATGCCTCCCGGGGAAGCCGTCGCTCGAAGGGACAACGGTCCAACCCGTAAACCAGCGCCATACACATG\n",
        "TGTCACGTAGGAGTACAAAAGTTCCACCATGCCAAAGCAACCCGTGGGACCAAGTGGAGGGGGCACAGAATACTG\n",
        "AAA\n",
        "\n",
        "\n",
        "\"\n",
        "\n",
        "\n",
        "target_gene = \n",
        "\n",
        "#the sequence of the gene of interest. This is the sequence of the gene\n",
        "#that was cloned into the plasmid. This is optional. If you don't need\n",
        "#a sequence analysis simply delete the example sequence below \n",
        "#but leave the quotation marks there. thx  \n",
        "  \n",
        "\"\n",
        "\n",
        "ATGTTTTGTTTCTTCCAAAACCGCTGGGAGTCGGGTCTTGGTGCCCATCTAACCGAGGAACTCATTAGTAGCGCATTCAG\n",
        "CGATTTTGCGCCTTACGCGGGGGCGGAGCAGCAACAGACACGTTTCGTCCGCCGCCAAAGCGATGGGACAACCACCAGAG\n",
        "TTGCTTATGGGGAAGATGGGGGGTGGTCGGTAAGCATGGGAAATCCTTACGCTGGCGGCATATCTTTTGCAACGACACCA\n",
        "GATTTACTCTGTGATATGCAACTGCGCATACCACTGAAGTTGGCTAGGGATGTTGTGGAGCAGAATAGCGAAAGGGGGAT\n",
        "CATTGGCAGTGCCGACAACTGTGAGGTGCTTCCAGTTTCCAGAATCTCCCTAACACCCATAACGACACCCTACGATAGTC\n",
        "GAAGTCCTTTTCAGTATTCTGTGCCCCCTCCACTTGGTCCCACGGGTTGCTTTGGCATGGTGGAACTTTTGTACTCCTAC\n",
        "GTGACACATGTGTATGGCGCTGGTTTACGGGTTGGACCGTTGTCCCTTCGAGCGACGGCTTCCCCGGGAGGCATGCCCTC\n",
        "AATAGACAATTCGTATGTGAGTGCTTTGGGCCAAACTGCCTGTCTCGTTGACGGTTCGTGGTGCGCTCCGTTGAGTGCCT\n",
        "TTCCGTCGTGTAACCTTACAGTGGCGCTGCAGCACCCCTTTATGGCCGTTGAGGGTGATACGCTTGCAGCAATTTGTTTG\n",
        "GAGAGTTCGGCGCCAAAAACACCAGAGGGACAAGGGTGCAATAATTTTCGTCTTCTTGTTGTGCAGACGCTTGCAACAAG\n",
        "TAATAGGCAAAACAAGCAGAATCCCAGTGATTCTCACTCGCATAGCTTCAGTTCGGTTCTCAGGGGCGACAAAAACTGGC\n",
        "GTGAGCTAGGCACACTCGTTGCTGTACAGAAATCATTTCGCAACCAACGCTTTCGACTCAGTGCTGCTGCTATGATACAC\n",
        "GACCGTGAATTAGACTATGAAACCGCAGCTTTGGCGGATGTAACAGCCATATTTTCTAGACCAACGGTACTGCGGTTGGG\n",
        "TTTCAACAGAGCCCGCCGTGTTGCCATGGGTATAGCGACAGAAATACTTAATGGACTTACCCTTACATTGGGTGTCCATT\n",
        "ACACAGCGGGGGACTTGCGGTTTGGACTTGAAGCCGCCCTTTAG\n",
        "\n",
        "\n",
        "\"\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title \n",
        "#packages\n",
        "require(stringr)\n",
        "install.packages(\"NameNeedle\")\n",
        "require(NameNeedle)\n",
        "#functions \n",
        "reverse_complement = function(seq)\n",
        "{ #takes ATGC characters but also\n",
        "  # R for A or G\n",
        "  # Y for C or T\n",
        "  # S for G or C\n",
        "  # W for A or T\n",
        "  # K for G or T\n",
        "  # M for A or C\n",
        "  # B for C, G, or T\n",
        "  # D for A, G, or T\n",
        "  # H for A, C, or T\n",
        "  # V for A, C, or G\n",
        "  # N for A, G, C, or T\n",
        "  \n",
        "  seq = unlist(strsplit(seq,\"\"))\n",
        "  \n",
        "  #compute complement\n",
        "  seq_complement = rep(NA,length(seq))\n",
        "  seq_complement[seq==\"A\"] = \"T\"\n",
        "  seq_complement[seq==\"T\"] = \"A\"\n",
        "  seq_complement[seq==\"C\"] = \"G\"\n",
        "  seq_complement[seq==\"G\"] = \"C\"\n",
        "  seq_complement[seq==\"R\"] = \"Y\"\n",
        "  seq_complement[seq==\"Y\"] = \"R\"\n",
        "  seq_complement[seq==\"S\"] = \"S\"\n",
        "  seq_complement[seq==\"W\"] = \"W\"\n",
        "  seq_complement[seq==\"K\"] = \"M\"\n",
        "  seq_complement[seq==\"M\"] = \"K\"\n",
        "  seq_complement[seq==\"B\"] = \"V\"\n",
        "  seq_complement[seq==\"V\"] = \"B\"\n",
        "  seq_complement[seq==\"D\"] = \"H\"\n",
        "  seq_complement[seq==\"H\"] = \"D\"\n",
        "  seq_complement[seq==\"N\"] = \"N\"\n",
        "  \n",
        "  #compute reverse of complement\n",
        "  seq_reverse_complement = rep(NA,length(seq))\n",
        "  seq_reverse_complement[length(seq):1] = seq_complement[1:length(seq)]\n",
        "  \n",
        "  return(paste(seq_reverse_complement,sep=\"\",collapse=\"\"))\n",
        "}\n",
        "\n",
        "\n",
        "translateDNAtoAA = function(seq)\n",
        "{ #an alogirthm for (fast) DNA to AA translation\n",
        "  #takes a string of ATGC DNA characters that are a\n",
        "  #open reading frame (ATG...STOP) and returns a \n",
        "  #string of amino acids\n",
        "\n",
        "  #packages\n",
        "  require(stringr)\n",
        "  \n",
        "  #get codons and unique codons\n",
        "  seq.tripplets = str_extract_all(seq,\"[ATGC]{3}\")[[1]]\n",
        "  seq.tripplets.unique = unique(seq.tripplets)\n",
        "  \n",
        "  { #load codon table\n",
        "    codons.1 = c(rep(\"T\",16),rep(\"C\",16),rep(\"A\",16),rep(\"G\",16))\n",
        "    codons.2 = rep(c(rep(\"T\",4),rep(\"C\",4),rep(\"A\",4),rep(\"G\",4)),4)\n",
        "    codons.3 = rep(c(\"T\",\"C\",\"A\",\"G\"),16)\n",
        "    one.letter.not = c(\"F\",\"F\",\"L\",\"L\",\"S\",\"S\",\"S\",\"S\",\"Y\",\"Y\",\"*\",\"*\",\"C\",\"C\",\"*\",\"W\",\n",
        "                       \"L\",\"L\",\"L\",\"L\",\"P\",\"P\",\"P\",\"P\",\"H\",\"H\",\"Q\",\"Q\",\"R\",\"R\",\"R\",\"R\",\n",
        "                       \"I\",\"I\",\"I\",\"M\",\"T\",\"T\",\"T\",\"T\",\"N\",\"N\",\"K\",\"K\",\"S\",\"S\",\"R\",\"R\",\n",
        "                       \"V\",\"V\",\"V\",\"V\",\"A\",\"A\",\"A\",\"A\",\"D\",\"D\",\"E\",\"E\",\"G\",\"G\",\"G\",\"G\")\n",
        "    codons = cbind(paste(codons.1,codons.2,codons.3,sep=\"\"),one.letter.not)\n",
        "  }#load codon table end\n",
        "\n",
        "\n",
        "  #translation to amino acids\n",
        "  aaString = rep(NA,length(seq.tripplets))\n",
        "  #find aa of every unique codon and combine to aa string\n",
        "  for (i in 1:length(seq.tripplets.unique)){\n",
        "    aa = codons[which(seq.tripplets.unique[i]==codons[,1]),2]\n",
        "    pos.of.codon = which(seq.tripplets==seq.tripplets.unique[i])\n",
        "    aaString[pos.of.codon] = aa\n",
        "  }\n",
        "  \n",
        "  return(paste(aaString,sep=\"\",collapse=\"\"))\n",
        "\n",
        "}\n",
        "\n",
        "\n",
        "\n",
        "#######################################################################################################\n",
        "\n",
        "sequencing_results_analyzer = function(seq1,seq2,target)\n",
        "{ #this algorithm takes sequencing results form a forward primer (20/32 on our plasmids)\n",
        "  #and a reverse primer (20/33 on our plasmids) which will overlapp in the middle.\n",
        "  #the two sequences are combined to the final sequence which is encoded on the plasmid\n",
        "  #and this plasmid is aligned against the gene that was cloned into the plasmid (target)\n",
        "  #to check for sequence errors.\n",
        "\n",
        "  #check if we have an RNAi plasmid\n",
        "  rnai=FALSE\n",
        "  if(str_detect(seq2,\"rnai\")){\n",
        "    rnai = TRUE\n",
        "    cat(\"Input sequences are from an RNAi plasmid and will be analyzed this way.\\n\")\n",
        "  }\n",
        "\n",
        "  seq_forward_primer = str_replace_all(seq1,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "  seq_reverse_primer = str_replace_all(seq2,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "\n",
        "  #break is the sequences are too short for analysis\n",
        "  if(nchar(seq_forward_primer)<101 | is.null(seq_forward_primer) | is.na(seq_forward_primer)) {\n",
        "    cat(\"The provided forward primer sequencing result is too short \\n The sequence is: \\n\",seq_forward_primer,)\n",
        "    break()\n",
        "    }\n",
        "  \n",
        "  if(!rnai)\n",
        "  { #non rnai plasmid analysis\n",
        "\n",
        "    #a revers primer result sequence is optional, check if it was provided \n",
        "    #if it was not provided, shortcut to aligments.\n",
        "    #if it was provided, compute combined sequence.\n",
        "    if(seq_reverse_primer==\"\") {sequence_combined = seq_forward_primer}\n",
        "    else{\n",
        "      #reverse complement of reverse primer result is needed\n",
        "      seq_reverse_primer = reverse_complement(seq_reverse_primer)\n",
        "      cat(\"the reverse complement of the sequencing result with the reverse primer is: \\n\",seq_reverse_primer,\"\\n\\n\")\n",
        "      \n",
        "      #scan for overlapp in the sequences between forward and reverse primer revcomp.\n",
        "      length_forward_primer = nchar(seq_forward_primer)\n",
        "      sequence_combined = NA\n",
        "      for(i in seq(round(length_forward_primer/2),length_forward_primer,length.out = 20)[-20])\n",
        "      { #the loop is simply to check 20 parts of the 2. half of forward primer result sequence \n",
        "        #for presence in the reverse primer result revcomp sequence\n",
        "        query = substr(seq_forward_primer,i,i+30) #value 30 randomely chosen by me\n",
        "        \n",
        "        if(str_detect(query,\"[NRYSWKMBDHV]\")){\n",
        "          #if sequence has sequencing error don't use it,\n",
        "          #go to next one\n",
        "          sequencing_error = TRUE\n",
        "          next()\n",
        "        }else{sequencing_error = FALSE}\n",
        "        \n",
        "        if(str_detect(seq_reverse_primer,query)){\n",
        "          sequence_combined = paste(substr(seq_forward_primer,1,i+30),\n",
        "                                    substr(seq_reverse_primer,str_locate(seq_reverse_primer,query)[2]+1,nchar(seq_reverse_primer)),\n",
        "                                    sep=\"\",collapse=\"\")\n",
        "          break()\n",
        "        }\n",
        "      }\n",
        "    }\n",
        "    if(!is.na(sequence_combined)){\n",
        "      #print the combined sequence encoded on the plasmid\n",
        "      cat(\"the sequence is: \\n\",sequence_combined)\n",
        "      cat(\"\\nthe sequence was generated with the first\",ifelse(seq_reverse_primer==\"\",nchar(seq_forward_primer),i+30),\n",
        "          \"nucleotides of the forward primer sequencing \\n\")\n",
        "      cat = ifelse(seq_reverse_primer==\"\",\"\\n\",paste(\"and the last\",nchar(seq_reverse_primer)-(str_locate(seq_reverse_primer,query)[2]),\"nucleotides of the reverse complement of the reverse primer sequencing\\n\\n\"))\n",
        "      cat(cat)\n",
        "    }else{\n",
        "      #if no combined sequence could be generated, report this together with\n",
        "      #help of what might be the problem.\n",
        "      if(sequencing_error){cat(\"no combined sequence could be generated\",\n",
        "                              \"sequencing errors were detected which could be the problem\\n\\n\")}\n",
        "      else{cat(\"please check input sequences, something unexpected went wrong.\\n\")}\n",
        "    }\n",
        "    \n",
        "    #analysis of the open reading frame in the combined sequence against \n",
        "    #the gene that was cloned into the plasmid. \n",
        "    #check if target gene was provided and prepare target sequence for alignment\n",
        "      target = str_replace_all(target,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "      if(target==\"\"){cat(\"\\n no target gene comparison can be made as no sequence\",\n",
        "                          \"\\n was provided. Job done.\");break()}\n",
        "    \n",
        "    #detect the open reading frame (OFR) and check it against a provided sequence\n",
        "    if(str_detect(sequence_combined,\"ATG([ATGC]{3})*((TGA)|(TAA)|(TAG))\"))\n",
        "    {\n",
        "      #detect open reading frame (ORF)\n",
        "      regex = \"ATG([ATGC]{3})+(?=((TGA)|(TAA)|(TAG)))\"\n",
        "      ORF = sequence_combined\n",
        "      while(str_detect(ORF,regex)){\n",
        "        #a while loop to detect the correct open reading frame\n",
        "        #starting at the first atg and ending at the closest inframe stop codon\n",
        "        ORF = str_extract(ORF,regex)\n",
        "        if(!str_detect(regex,\"\\\\^\")) {regex = paste0(\"^\",regex)}\n",
        "      }\n",
        "      cat(\"an open reading frame was detected, the sequence is:\\n\",ORF,\"\\n\\n\")\n",
        "      \n",
        "      #change parameters for Needleman Wunsch algorithm\n",
        "      myParam <- defaultNeedleParams\n",
        "      myParam$GAPCHAR = \"-\"\n",
        "\n",
        "      alignment = needles(pattern = target,subject = ORF,params = myParam)\n",
        "\n",
        "      #needles from NameNeedle is bugging around in some local maxima/minima\n",
        "      #especially when one seqeunce is much longer than the other one.\n",
        "      #as a result it will introduce unlogical frame shifts in the end of the \n",
        "      #short sequence to expand its length. This is annoying and I cannot fix\n",
        "      #it because i cannot punish many short more than one long frameshift.\n",
        "      #therefor, I will fix this by removing all these downstream frame shifts\n",
        "      #manually and hope that I don't remove an actual one, although this would\n",
        "      #still be clearly visible in the output...\n",
        "      if(nchar(target) < nchar(ORF)){\n",
        "        #remove indels in pattern\n",
        "        alignment$align1 = paste(substr(alignment$align1,1,str_locate(alignment$align1,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2] -1),\n",
        "                                    str_replace_all(substr(alignment$align1,str_locate(alignment$align1,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2],nchar(alignment$align1)),\"-\",\"\"),\n",
        "                                    sep=\"\",collapse=\"\") #ugly af but works\n",
        "                                    #\".*[ATGCNRYSWKMBDHV]{30,}[-]\" is a regex to find the first downstream frame shift induced by the bug.\n",
        "\n",
        "\n",
        "        #print this a little nicer into the console\n",
        "        pattern = unlist(strsplit(as.character(alignment$align1),\"\"))\n",
        "        subject = unlist(strsplit(as.character(alignment$align2),\"\"))\n",
        "\n",
        "        #some cosmetics\n",
        "        pattern = c(pattern,rep(\"-\",length(subject)-length(pattern)))\n",
        "        }\n",
        "      if(nchar(target) > nchar(ORF)){\n",
        "        #remove indels in subject\n",
        "        alignment$align2 = paste(substr(alignment$align2,1,str_locate(alignment$align2,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2] -1),\n",
        "                                    str_replace_all(substr(alignment$align2,str_locate(alignment$align2,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2],nchar(alignment$align2)),\"-\",\"\"),\n",
        "                                    sep=\"\",collapse=\"\") #ugly af but works\n",
        "                                    #\".*[ATGCNRYSWKMBDHV]{30,}[-]\" is a regex to find the first downstream frame shift induced by the bug.\n",
        "\n",
        "\n",
        "        #print this a little nicer into the console\n",
        "        pattern = unlist(strsplit(as.character(alignment$align1),\"\"))\n",
        "        subject = unlist(strsplit(as.character(alignment$align2),\"\"))\n",
        "\n",
        "        #some cosmetics\n",
        "        subject = c(subject,rep(\"-\",length(pattern)-length(subject)))\n",
        "        }\n",
        "      if(nchar(target) == nchar(ORF)){\n",
        "        #no indels have to be removed.\n",
        "        #print this a little nicer into the console\n",
        "        pattern = unlist(strsplit(as.character(alignment$align1),\"\"))\n",
        "        subject = unlist(strsplit(as.character(alignment$align2),\"\"))\n",
        "\n",
        "        }\n",
        "      \n",
        "\n",
        "      matching = rep(\" \",length(pattern))\n",
        "      #prepare to print into console\n",
        "      matching[which(pattern==subject)] = \"|\"\n",
        "      matching = paste(matching,sep=\"\",collapse=\"\")\n",
        "      pattern = paste(pattern,sep=\"\",collapse=\"\")\n",
        "\n",
        "      alignment.out = as.character(c(pattern, matching, alignment$align2))\n",
        "\n",
        "      cat(\"the ORF sequence alignment to the target sequence looks as follows:\\n\",\n",
        "      \"(1: target gene, 2: matching/not matching, 3: open reading frame obtained from sequencing)\\n\")\n",
        "      l = 1\n",
        "      continue.print = TRUE\n",
        "      while(continue.print){\n",
        "        if(l==nchar(pattern)){continue.print = FALSE}\n",
        "        else{pattern.out = substr(pattern,l,l+99)\n",
        "        matching.out = substr(matching,l,l+99)\n",
        "        alignment.out = substr(as.character(alignment$align2),l,l+99)\n",
        "        print(c(pattern.out,matching.out,alignment.out))\n",
        "        l = l+100\n",
        "        l = ifelse(l<=nchar(pattern),l,nchar(pattern))\n",
        "        cat(\"\\n\")\n",
        "        }\n",
        "      }\n",
        "      \n",
        "\n",
        "      #translate ORF and target gene to protein (AA) and compare sequences\n",
        "      if(!str_detect(ORF,\"[^ATGC]\")){\n",
        "      ORF_translate = translateDNAtoAA(ORF)\n",
        "      }else{cat(\"\\nsequencing results contain unread bases and cannot be translated.\\n\");break()}\n",
        "      cat(\"\\n\\n the translated open reading frame is:\\n\",ORF_translate)\n",
        "\n",
        "      HA_tag_count = str_count(ORF_translate,\"YPYDVPDYA\")\n",
        "      myc_tag_count = str_count(ORF_translate,\"EQKLISEEDL\")\n",
        "      TEV_site_count = str_count(ORF_translate,\"ENLYFQGS\")\n",
        "\n",
        "      cat(\"\\n\\nthe translated sequence contains\",HA_tag_count,\"HA tags ;\",myc_tag_count,\"myc tags ; \",TEV_site_count,\"TEV cleavage sites.\")\n",
        "\n",
        "      #detect open reading frame in target gene and translate it (only if all bases are ATGC)\n",
        "      if(str_detect(target,\"ATG([ATGC]{3})*((TGA)|(TAA)|(TAG))\") & !str_detect(target,\"[^ATGC]\"))\n",
        "      {\n",
        "      #detect open reading frame (ORF)\n",
        "      regex = \"ATG([ATGC]{3})+(?=((TGA)|(TAA)|(TAG)))\"\n",
        "      ORF_target = target\n",
        "      while(str_detect(ORF_target,regex)){\n",
        "        #a while loop to detect the correct open reading frame\n",
        "        #starting at the first atg and ending at the closest inframe stop codon\n",
        "        ORF_target = str_extract(ORF_target,regex)\n",
        "        if(!str_detect(regex,\"\\\\^\")) {regex = paste0(\"^\",regex)}\n",
        "      }\n",
        "      }else{cat(\"\\nprovided target gene does not contain an open reading frame\",\n",
        "                  \"\\nand/or unread bases and thus cannot be translated.\\n\");break()}\n",
        "\n",
        "      #translate the open reading frame of the target sequence\n",
        "      ORF_target_translate = translateDNAtoAA(ORF_target)\n",
        "\n",
        "      #compare the translated sequences using the Needleman Wunsch algorithm\n",
        "\n",
        "      alignment_aa = needles(pattern = ORF_target_translate, subject = ORF_translate, params = myParam)\n",
        "\n",
        "      #needles from NameNeedle is bugging around in some local maxima/minima\n",
        "      #especially when one seqeunce is much longer than the other one.\n",
        "      #as a result it will introduce unlogical frame shifts in the end of the \n",
        "      #short sequence to expand its length. This is annoying and I cannot fix\n",
        "      #it because i cannot punish many short more than one long frameshift.\n",
        "      #therefor, I will fix this by removing all these downstream frame shifts\n",
        "      #manually and hope that I don't remove an actual one, although this would\n",
        "      #still be clearly visible in the output...\n",
        "      if(nchar(ORF_target_translate) < nchar(ORF_translate)){\n",
        "        #remove indels in pattern\n",
        "        alignment_aa$align1 = paste(substr(alignment_aa$align1,1,str_locate(alignment_aa$align1,\".*[A-Z]{10,}[-]\")[2] -1),\n",
        "                                    str_replace_all(substr(alignment_aa$align1,str_locate(alignment_aa$align1,\".*[A-Z]{10,}[-]\")[2],nchar(alignment_aa$align1)),\"-\",\"\"),\n",
        "                                    sep=\"\",collapse=\"\") #ugly af but works\n",
        "                                    #\".*[A-Z]{10,}[-]\" is a regex to find the first downstream frame shift induced by the bug.\n",
        "\n",
        "\n",
        "        #print this a little nicer into the console\n",
        "        pattern_aa = unlist(strsplit(as.character(alignment_aa$align1),\"\"))\n",
        "        subject_aa = unlist(strsplit(as.character(alignment_aa$align2),\"\"))\n",
        "\n",
        "        #some cosmetics\n",
        "        pattern_aa = c(pattern_aa,rep(\"-\",length(subject_aa)-length(pattern_aa)))\n",
        "      }\n",
        "\n",
        "      if(nchar(ORF_target_translate) > nchar(ORF_translate)){\n",
        "        #remove indels in subject\n",
        "        alignment_aa$align2 = paste(substr(alignment_aa$align2,1,str_locate(alignment_aa$align2,\".*[A-Z]{10,}[-]\")[2] -1),\n",
        "                                    str_replace_all(substr(alignment_aa$align2,str_locate(alignment_aa$align2,\".*[A-Z]{10,}[-]\")[2],nchar(alignment_aa$align2)),\"-\",\"\"),\n",
        "                                    sep=\"\",collapse=\"\") #ugly af but works\n",
        "                                    #\".*[A-Z]{10,}[-]\" is a regex to find the first downstream frame shift induced by the bug.\n",
        "\n",
        "\n",
        "        #print this a little nicer into the console\n",
        "        pattern_aa = unlist(strsplit(as.character(alignment_aa$align1),\"\"))\n",
        "        subject_aa = unlist(strsplit(as.character(alignment_aa$align2),\"\"))\n",
        "\n",
        "        #some cosmetics\n",
        "        subject_aa = c(subject_aa,rep(\"-\",length(pattern_aa)-length(subject_aa)))\n",
        "      }\n",
        "\n",
        "      if(nchar(ORF_target_translate) > nchar(ORF_translate)){\n",
        "        #nothing has to be changed\n",
        "        #print this a little nicer into the console\n",
        "        pattern_aa = unlist(strsplit(as.character(alignment_aa$align1),\"\"))\n",
        "        subject_aa = unlist(strsplit(as.character(alignment_aa$align2),\"\"))\n",
        "      }\n",
        "\n",
        "      matching_aa = rep(\" \",length(pattern_aa))\n",
        "      #prepare to print into console\n",
        "      matching_aa[which(pattern_aa==subject_aa)] = \"|\"\n",
        "      matching_aa = paste(matching_aa,sep=\"\",collapse=\"\")\n",
        "      pattern_aa = paste(pattern_aa,sep=\"\",collapse=\"\")\n",
        "\n",
        "      alignment.out_aa = as.character(c(pattern_aa, matching_aa, alignment_aa$align2))\n",
        "\n",
        "      cat(\"\\n\\nthe translated ORF sequence alignment to the translated target sequence looks as follows:\\n\",\n",
        "      \"(1: target gene aa, 2: matching/not matching, 3: translated open reading frame obtained from sequencing)\\n\")\n",
        "\n",
        "      l = 1\n",
        "      continue.print = TRUE\n",
        "      while(continue.print){\n",
        "        if(l==nchar(pattern_aa)){continue.print = FALSE}\n",
        "        else{pattern_aa.out = substr(pattern_aa,l,l+99)\n",
        "        matching_aa.out = substr(matching_aa,l,l+99)\n",
        "        alignment_aa.out = substr(as.character(alignment_aa$align2),l,l+99)\n",
        "        print(c(pattern_aa.out,matching_aa.out,alignment_aa.out))\n",
        "        l = l+100\n",
        "        l = ifelse(l<=nchar(pattern_aa),l,nchar(pattern_aa))\n",
        "        cat(\"\\n\")\n",
        "        }\n",
        "      }\n",
        "\n",
        "\n",
        "    }else{cat(\"No open reading frame could be detected in the sequence.\\n\")}\n",
        "  }\n",
        "\n",
        "\n",
        "  if(rnai)\n",
        "  { #analysis of RNAi plasmids is straight forward:\n",
        "    #make sequence alignment using Needleman Wunsch alogrithm\n",
        "    #of both inserts to target sequence\n",
        "\n",
        "    #check if target gene was provided and prepare target sequence for alignment\n",
        "      target = str_replace_all(target,\"[^ATGCNRYSWKMBDHV]\",\"\")\n",
        "      if(target==\"\"){cat(\"\\n no target gene comparison can be made as no sequence\",\n",
        "                          \"\\n was provided. Job done.\");break()}\n",
        "\n",
        "    #change parameters for Needleman Wunsch algorithm\n",
        "      myParam <- defaultNeedleParams\n",
        "      myParam$GAPCHAR = \"-\"\n",
        "\n",
        "    #forward insert\n",
        "      alignment1 = needles(pattern = target,subject = seq_forward_primer,params = myParam)\n",
        "\n",
        "      #needles from NameNeedle is bugging around in some local maxima/minima\n",
        "      #especially when one seqeunce is much longer than the other one.\n",
        "      #as a result it will introduce unlogical frame shifts in the end of the \n",
        "      #short sequence to expand its length. This is annoying and I cannot fix\n",
        "      #it because i cannot punish many short more than one long frameshift.\n",
        "      #therefor, I will fix this by removing all these downstream frame shifts\n",
        "      #manually and hope that I don't remove an actual one, although this would\n",
        "      #still be clearly visible in the output...\n",
        "      #we only do this in the pattern sequence because this (in RNAi plamids)\n",
        "      #will always be a sequence <600 bp whereas sequencing reactions usually \n",
        "      #give around 1000 bp.\n",
        "      alignment1$align1 = paste(substr(alignment1$align1,1,str_locate(alignment1$align1,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2] -1),\n",
        "                                  str_replace_all(substr(alignment1$align1,str_locate(alignment1$align1,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2],nchar(alignment1$align1)),\"-\",\"\"),\n",
        "                                  sep=\"\",collapse=\"\") #ugly af but works\n",
        "                                  #\".*[ATGCNRYSWKMBDHV]{30,}[-]\" is a regex to find the first downstream frame shift induced by the bug.\n",
        "\n",
        "\n",
        "      #print this a little nicer into the console\n",
        "      pattern1 = unlist(strsplit(as.character(alignment1$align1),\"\"))\n",
        "      subject1 = unlist(strsplit(as.character(alignment1$align2),\"\"))\n",
        "\n",
        "      #some cosmetics\n",
        "      pattern1 = c(pattern1,rep(\"-\",length(subject1)-length(pattern1)))\n",
        "\n",
        "      matching1 = rep(\" \",length(pattern1))\n",
        "\n",
        "      #prepare to print into console\n",
        "      matching1[which(pattern1==subject1)] = \"|\"\n",
        "      matching1 = paste(matching1,sep=\"\",collapse=\"\")\n",
        "      pattern1 = paste(pattern1,sep=\"\",collapse=\"\")\n",
        "\n",
        "      alignment.out.1 = as.character(c(pattern1, matching1, alignment1$align2))\n",
        "\n",
        "      cat(\"\\nthe sequence alignment of the forward insert to the target sequence looks as follows:\\n\",\n",
        "      \"(1: target gene, 2: matching/not matching, 3: RNAi forward insert (primer: 20/32))\\n\")\n",
        "\n",
        "      l = 1\n",
        "      continue.print = TRUE\n",
        "      while(continue.print){\n",
        "        if(l==nchar(pattern1)){continue.print = FALSE}\n",
        "        else{pattern1.out = substr(pattern1,l,l+99)\n",
        "        matching1.out = substr(matching1,l,l+99)\n",
        "        alignment1.out = substr(as.character(alignment1$align2),l,l+99)\n",
        "        print(c(pattern1.out,matching1.out,alignment1.out))\n",
        "        l = l+100\n",
        "        l = ifelse(l<=nchar(pattern1),l,nchar(pattern1))\n",
        "        cat(\"\\n\")\n",
        "        }\n",
        "      }\n",
        "\n",
        "\n",
        "    #reverse insert\n",
        "      alignment2 = needles(pattern = target,subject = seq_reverse_primer,params = myParam)\n",
        "      #the sequence must not be the reverse complement, as the insert has to be reversed!\n",
        "\n",
        "      #the algorithm introduces lots of frame shifts in the end of the pattern,\n",
        "      #because this is short than the sequencing result. So we remove them by hand\n",
        "      alignment2$align1 = paste(substr(alignment2$align1,1,str_locate(alignment2$align1,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2] -1),\n",
        "                                  str_replace_all(substr(alignment2$align1,str_locate(alignment2$align1,\".*[ATGCNRYSWKMBDHV]{30,}[-]\")[2],nchar(alignment2$align1)),\"-\",\"\"),\n",
        "                                  sep=\"\",collapse=\"\") #ugly af but works\n",
        "                                  #\".*[ATGCNRYSWKMBDHV]{30,}[-]\" is a regex to find the first downstream frame shift induced by the bug.\n",
        "\n",
        "      #print this a little nicer into the console\n",
        "      pattern2 = unlist(strsplit(as.character(alignment2$align1),\"\"))\n",
        "      subject2 = unlist(strsplit(as.character(alignment2$align2),\"\"))\n",
        "\n",
        "      #some cosmetics\n",
        "      pattern2 = c(pattern2,rep(\"-\",length(subject2)-length(pattern2)))\n",
        "\n",
        "      matching2 = rep(\" \",length(pattern2))\n",
        "\n",
        "      #prepare to print into console\n",
        "      matching2[which(pattern2==subject2)] = \"|\"\n",
        "      matching2 = paste(matching2,sep=\"\",collapse=\"\")\n",
        "      pattern2 = paste(pattern2,sep=\"\",collapse=\"\")\n",
        "\n",
        "      alignment.out.2 = as.character(c(pattern2, matching2, alignment2$align2))\n",
        "\n",
        "      cat(\"\\n\\nthe sequence alignment of the reverse insert to the target sequence looks as follows:\\n\",\n",
        "      \"(1: target gene, 2: matching/not matching, 3: RNAi reverse insert (primer: 20/33))\\n\")\n",
        "\n",
        "        l = 1\n",
        "      continue.print = TRUE\n",
        "      while(continue.print){\n",
        "        if(l==nchar(pattern2)){continue.print = FALSE}\n",
        "        else{pattern2.out = substr(pattern2,l,l+99)\n",
        "        matching2.out = substr(matching2,l,l+99)\n",
        "        alignment2.out = substr(as.character(alignment2$align2),l,l+99)\n",
        "        print(c(pattern2.out,matching2.out,alignment2.out))\n",
        "        l = l+100\n",
        "        l = ifelse(l<=nchar(pattern2),l,nchar(pattern2))\n",
        "        cat(\"\\n\")\n",
        "        }\n",
        "      }\n",
        "  }\n",
        "\n",
        "\n",
        "}\n",
        "\n",
        "sequencing_results_analyzer(seq1 = sequence_forward_primer,seq2 = sequence_reverse_primer,target = target_gene)\n",
        "\n"
      ],
      "metadata": {
        "id": "wD_Y3H525M4d",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}